sqtls = {
    '/variants/genes': [{
        'params': {
            'p_value': 'lte:0.01',
            'label': 'splice_QTL',
            'verbose': 'false'
        },
        'response': [
            {
                'sequence variant': 'variants/6972da7e0959b2b1a747762684cf9c431ab75c43b32cce71c8c53e012a486cb6',
                'gene': 'genes/ENSG00000188290',
                'p_value': 0.00000213399,
                'slope': 0.497641,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/4f84322a236475966feb68b2e5589c2c51b2c226604f77f59f152e1dbd2c5648',
                'gene': 'genes/ENSG00000188290',
                'p_value': 1.09941e-11,
                'slope': -0.683346,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/1323d67465672ea0ca3c25cf95b76b7c78aca52a2379416630a9544df5d17a93',
                'gene': 'genes/ENSG00000188290',
                'p_value': 1.36453e-13,
                'slope': -0.735844,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/f50d42c0bf357fe71fbd0f5ddcd73efbcd48efe10ae2189e28a578d58036aaff',
                'gene': 'genes/ENSG00000188290',
                'p_value': 7.81183e-12,
                'slope': -0.68954,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/362436e9df3d27ccbbb9713cde1fc5e0e896446113822a011a8804bfe3a35f7d',
                'gene': 'genes/ENSG00000188290',
                'p_value': 4.87218e-14,
                'slope': -0.743714,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/22f170e54c30a59e737beba20444f192201126f0b1415a7c9a106d1d01fe98d0',
                'gene': 'genes/ENSG00000188290',
                'p_value': 2.06171e-11,
                'slope': -0.814103,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/0e82ef1610de3c07a4b9a57ac0ccf090da31cd1bba048a25b5680c41f950dbf5',
                'gene': 'genes/ENSG00000188290',
                'p_value': 1.18979e-15,
                'slope': -0.781824,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/85056aaee88e687530ffafe66566942b2ea4c97cc1e9ad6b8d3316e0c2ca03d7',
                'gene': 'genes/ENSG00000188290',
                'p_value': 6.63382e-9,
                'slope': 0.724306,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/2c709a08569f57d8532d95ee7a70082114fa12491f984cb8f336cf90bc18323f',
                'gene': 'genes/ENSG00000188290',
                'p_value': 6.63382e-9,
                'slope': 0.724306,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/2cb8de7458336a8afcacca1ecb09be6bde9530710664af37d1b50e6064100a50',
                'gene': 'genes/ENSG00000188290',
                'p_value': 8.47956e-16,
                'slope': -0.801873,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/a769b391bc07d5f76c3a65005a506404cc976a34ece96795aa7d51c7586cb5f7',
                'gene': 'genes/ENSG00000188290',
                'p_value': 2.30621e-16,
                'slope': -0.807764,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/caf4b2f52b12eac647f51cd9adcadf7aec2af4f0a4ed974a3e91a5c9d2f8cd3f',
                'gene': 'genes/ENSG00000188290',
                'p_value': 9.27251e-8,
                'slope': 0.660596,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/b1ea03667bd5a56cc23bf36f0917a38c8616efab8419ed5f00b112d5a5e5d2d8',
                'gene': 'genes/ENSG00000188290',
                'p_value': 4.94533e-10,
                'slope': -0.601093,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/2bafef001b1e7bb6c12eb79ec305a147a4b5ea8c08e5aeb20dbb6b7b60b52d25',
                'gene': 'genes/ENSG00000188290',
                'p_value': 1.79152e-8,
                'slope': 0.811672,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/5ce175e66a833b4d3fd69d977ec5c6e0b73d45937f1574e656e244b74e4224eb',
                'gene': 'genes/ENSG00000188290',
                'p_value': 9.94503e-8,
                'slope': 0.741419,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/4298328c6b48b658d3b054d31b8cc917e47da077504a2437d2a5fb9beb0fd561',
                'gene': 'genes/ENSG00000188290',
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/e15f3a4c5e3fbb33bc0f54b168e16208013dc5b0225392caa123679c4a1eb90e',
                'gene': 'genes/ENSG00000188290',
                'p_value': 6.47548e-7,
                'slope': 0.634734,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/ff45fb4d95bd827f73ec7d08cb2bd28c5bcec222e93bdd2d73009378e34a2997',
                'gene': 'genes/ENSG00000188290',
                'p_value': 3.66638e-23,
                'slope': -0.946632,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/f24a1e2049bfd5375f5609a52903ae3176d071fbdf33c9dfd45e330eb4606211',
                'gene': 'genes/ENSG00000188290',
                'p_value': 6.47548e-7,
                'slope': 0.634734,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/9386dd3a5b27f4631a7ad5e464d2e64947ebc71d241c73ffc34251f9940ea231',
                'gene': 'genes/ENSG00000188290',
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/14e9c0b218f3cbd186fe91800d24caf4bd77459110d2897f45be6da709408cc6',
                'gene': 'genes/ENSG00000188290',
                'p_value': 2.03398e-22,
                'slope': -0.899991,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/75a7f16eea568ee3fb51bfa17d5880c22679ef21986e9a9812a290ee0d994912',
                'gene': 'genes/ENSG00000188290',
                'p_value': 1.27736e-20,
                'slope': -0.925848,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/d2951bddb245eba65291f34d65c487c180940a923a5f93769ed3c06d514966c5',
                'gene': 'genes/ENSG00000188290',
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/016b278013174d0a810967181cd46decdf6d040c0a97a5e096d4983a192b3f5b',
                'gene': 'genes/ENSG00000188290',
                'p_value': 3.89225e-7,
                'slope': 0.644408,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': 'variants/3be794b104906c9257577eb766f68df91302a58f380c6f2b14e40a8bdfcf31e3',
                'gene': 'genes/ENSG00000188290',
                'p_value': 2.3738e-24,
                'slope': -0.928406,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            }
        ]
    }, {
        'params': {
            'p_value': 'lte:0.01',
            'label': 'splice_QTL',
            'verbose': 'true'
        },
        'response': [
            {
                'sequence variant': {
                    '_key': '6972da7e0959b2b1a747762684cf9c431ab75c43b32cce71c8c53e012a486cb6',
                    '_id': 'variants/6972da7e0959b2b1a747762684cf9c431ab75c43b32cce71c8c53e012a486cb6',
                    '_rev': '_geyXI0y--e',
                    'chr': 'chr1',
                    'pos:long': 971789,
                    'rsid': [
                        'rs113797882'
                    ],
                    'ref': 'AG',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.6979,
                                'alt:long': 0.3021
                            },
                            'alspac': {
                                'ref:long': 0.6196,
                                'alt:long': 0.3804
                            },
                            'estonian': {
                                'ref:long': 0.619,
                                'alt:long': 0.381
                            },
                            'genome_dk': {
                                'ref:long': 0.6,
                                'alt:long': 0.4
                            },
                            'gnomad': {
                                'ref:long': 0.6531,
                                'alt:long': 0.3469
                            },
                            'gonl': {
                                'ref:long': 0.5932,
                                'alt:long': 0.4068
                            },
                            'korea1k': {
                                'ref:long': 0.8392,
                                'alt:long': 0.1608
                            },
                            'northernsweden': {
                                'ref:long': 0.5983,
                                'alt:long': 0.4017
                            },
                            'tommo': {
                                'ref:long': 0.8293,
                                'alt:long': 0.1707
                            },
                            'topmed': {
                                'ref:long': 0.6657,
                                'alt:long': 0.3343
                            },
                            'twinsuk': {
                                'ref:long': 0.6343,
                                'alt:long': 0.3657
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.6401,
                                'alt:long': 0.3599
                            }
                        },
                        'varinfo': '1-971790-AG-A',
                        'vid': '8.81294e+09',
                        'variant_vcf': '1-971790-AG-A',
                        'variant_annovar': '1-971791-971791-G--',
                        'start_position': '971791',
                        'end_position': '971791',
                        'ref_annovar': 'G',
                        'alt_annovar': '-',
                        'ref_vcf': 'AG',
                        'alt_vcf': 'A',
                        'bravo_an': '264690',
                        'bravo_af': '0.334255',
                        'filter_status': 'PASS',
                        'af_total': 0.347152,
                        'af_asj_female': 0.342045,
                        'af_eas_female': 0.160912,
                        'af_afr_male': 0.30029,
                        'af_female': 0.348265,
                        'af_fin_male': 0.427421,
                        'af_oth_female': 0.346715,
                        'af_ami': 0.459911,
                        'af_oth': 0.353953,
                        'af_male': 0.345969,
                        'af_ami_female': 0.465812,
                        'af_afr': 0.301563,
                        'af_eas_male': 0.143876,
                        'af_sas': 0.359816,
                        'af_nfe_female': 0.380663,
                        'af_asj_male': 0.334187,
                        'af_raw': 0.347267,
                        'af_oth_male': 0.36148,
                        'af_nfe_male': 0.379543,
                        'af_asj': 0.33835,
                        'af_amr_male': 0.294011,
                        'af_amr_female': 0.317321,
                        'af_fin': 0.429214,
                        'af_afr_female': 0.302648,
                        'af_sas_male': 0.354956,
                        'af_amr': 0.3041,
                        'af_nfe': 0.380192,
                        'af_eas': 0.151757,
                        'af_ami_male': 0.453488,
                        'af_fin_female': 0.434904,
                        'cadd_rawscore': '-0.090108',
                        'cadd_phred': '1.065',
                        'tg_afr': '0.2935',
                        'tg_all': '0.302117',
                        'tg_amr': '0.3213',
                        'tg_eas': '0.1567',
                        'tg_eur': '0.3618',
                        'tg_sas': '0.3885'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 0.00000213399,
                'slope': 0.497641,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '4f84322a236475966feb68b2e5589c2c51b2c226604f77f59f152e1dbd2c5648',
                    '_id': 'variants/4f84322a236475966feb68b2e5589c2c51b2c226604f77f59f152e1dbd2c5648',
                    '_rev': '_geyXI6S--e',
                    'chr': 'chr1',
                    'pos:long': 978508,
                    'rsid': [
                        'rs2340596'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.48,
                                'alt:long': 0.52
                            },
                            'alspac': {
                                'ref:long': 0.3884,
                                'alt:long': 0.6116
                            },
                            'estonian': {
                                'ref:long': 0.4585,
                                'alt:long': 0.5415
                            },
                            'genome_dk': {
                                'ref:long': 0.45,
                                'alt:long': 0.55
                            },
                            'gnomad': {
                                'ref:long': 0.5009,
                                'alt:long': 0.4991
                            },
                            'korean': {
                                'ref:long': 0.2532,
                                'alt:long': 0.7468
                            },
                            'northernsweden': {
                                'ref:long': 0.42,
                                'alt:long': 0.58
                            },
                            'page_study': {
                                'ref:long': 0.5163,
                                'alt:long': 0.4837
                            },
                            'qatari': {
                                'ref:long': 0.6528,
                                'alt:long': 0.3472
                            },
                            'sgdp_prj': {
                                'ref:long': 0.258,
                                'alt:long': 0.742
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0.7273
                            },
                            'tommo': {
                                'ref:long': 0.2503,
                                'alt:long': 0.7497
                            },
                            'topmed': {
                                'ref:long': 0.4931,
                                'alt:long': 0.5069
                            },
                            'twinsuk': {
                                'ref:long': 0.3897,
                                'alt:long': 0.6103
                            },
                            'vietnamese': {
                                'ref:long': 0.2897,
                                'alt:long': 0.7103
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4514,
                                'alt:long': 0.5486
                            }
                        },
                        'varinfo': '1-978509-G-A',
                        'vid': '2.45552e+06',
                        'variant_vcf': '1-978509-G-A',
                        'variant_annovar': '1-978509-978509-G-A',
                        'start_position': '978509',
                        'end_position': '978509',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 0.0454993,
                        'apc_conservation_v2': 0.0885926,
                        'apc_epigenetics_active': 9.90666,
                        'apc_epigenetics': 11.3248,
                        'apc_epigenetics_repressed': 4.77547,
                        'apc_epigenetics_transcription': 2.53859,
                        'apc_local_nucleotide_diversity': 2.72031,
                        'apc_local_nucleotide_diversity_v2': 0.206695,
                        'apc_local_nucleotide_diversity_v3': 0.209121,
                        'apc_mappability': 12.6226,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.207276,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 1.1291,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 9.20271,
                        'apc_transcription_factor': 15.8949,
                        'bravo_an': '264690',
                        'bravo_af': '0.506902',
                        'filter_status': 'PASS',
                        'fathmm_xf': '1.32427',
                        'funseq_description': 'noncoding',
                        'af_total': 0.501516,
                        'af_asj_female': 0.567537,
                        'af_eas_female': 0.718923,
                        'af_afr_male': 0.287557,
                        'af_female': 0.499661,
                        'af_fin_male': 0.521368,
                        'af_oth_female': 0.516393,
                        'af_ami': 0.423937,
                        'af_oth': 0.518622,
                        'af_male': 0.50349,
                        'af_ami_female': 0.412393,
                        'af_afr': 0.291353,
                        'af_eas_male': 0.726786,
                        'af_sas': 0.611877,
                        'af_nfe_female': 0.590427,
                        'af_asj_male': 0.584082,
                        'af_raw': 0.501472,
                        'af_oth_male': 0.520952,
                        'af_nfe_male': 0.583192,
                        'af_asj': 0.575301,
                        'af_amr_male': 0.644499,
                        'af_amr_female': 0.620468,
                        'af_fin': 0.523805,
                        'af_afr_female': 0.294585,
                        'af_sas_male': 0.616345,
                        'af_amr': 0.634105,
                        'af_nfe': 0.587383,
                        'af_eas': 0.723146,
                        'af_ami_male': 0.43662,
                        'af_fin_female': 0.53155,
                        'cadd_rawscore': '-0.501527',
                        'cadd_phred': '0.053',
                        'refseq_category': 'intronic',
                        'tg_afr': '0.208',
                        'tg_all': '0.519968',
                        'tg_amr': '0.6182',
                        'tg_eas': '0.7113',
                        'tg_eur': '0.6004',
                        'tg_sas': '0.592'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 1.09941e-11,
                'slope': -0.683346,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '1323d67465672ea0ca3c25cf95b76b7c78aca52a2379416630a9544df5d17a93',
                    '_id': 'variants/1323d67465672ea0ca3c25cf95b76b7c78aca52a2379416630a9544df5d17a93',
                    '_rev': '_geyXI6q--E',
                    'chr': 'chr1',
                    'pos:long': 978952,
                    'rsid': [
                        'rs13302979'
                    ],
                    'ref': 'C',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4437,
                                'alt:long': 0.5563
                            },
                            'alspac': {
                                'ref:long': 0.3861,
                                'alt:long': 0.6139
                            },
                            'estonian': {
                                'ref:long': 0.4433,
                                'alt:long': 0.5567
                            },
                            'exac': {
                                'ref:long': 0.3383,
                                'alt:long': 0.6617
                            },
                            'genome_dk': {
                                'ref:long': 0.45,
                                'alt:long': 0.55
                            },
                            'gnomad': {
                                'ref:long': 0.4787,
                                'alt:long': 0.5213
                            },
                            'gnomad_exomes': {
                                'ref:long': 0.3831,
                                'alt:long': 0.6169
                            },
                            'korean': {
                                'ref:long': 0.1733,
                                'alt:long': 0.8267
                            },
                            'northernsweden': {
                                'ref:long': 0.42,
                                'alt:long': 0.58
                            },
                            'qatari': {
                                'ref:long': 0.6111,
                                'alt:long': 0.3889
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2415,
                                'alt:long': 0.7585
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0.7273
                            },
                            'tommo': {
                                'ref:long': 0.1738,
                                'alt:long': 0.8262
                            },
                            'topmed': {
                                'ref:long': 0.4695,
                                'alt:long': 0.5305
                            },
                            'twinsuk': {
                                'ref:long': 0.3865,
                                'alt:long': 0.6135
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4399,
                                'alt:long': 0.5601
                            }
                        },
                        'varinfo': '1-978953-C-G',
                        'vid': '2.45686e+06',
                        'variant_vcf': '1-978953-C-G',
                        'variant_annovar': '1-978953-978953-C-G',
                        'start_position': '978953',
                        'end_position': '978953',
                        'ref_annovar': 'C',
                        'alt_annovar': 'G',
                        'ref_vcf': 'C',
                        'alt_vcf': 'G',
                        'aloft_value': '0.40355',
                        'aloft_description': 'Tolerant',
                        'apc_conservation': 2.02367,
                        'apc_conservation_v2': 3.98454,
                        'apc_epigenetics_active': 4.11315,
                        'apc_epigenetics': 7.03848,
                        'apc_epigenetics_repressed': 3.45095,
                        'apc_epigenetics_transcription': 1.16296,
                        'apc_local_nucleotide_diversity': 2.70482,
                        'apc_local_nucleotide_diversity_v2': 0.0457027,
                        'apc_local_nucleotide_diversity_v3': 0.0562811,
                        'apc_mappability': 3.75571,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0454695,
                        'apc_protein_function': 20.7948,
                        'apc_protein_function_v2': 20.7948,
                        'apc_protein_function_v3': 20.7948,
                        'apc_proximity_to_coding': 26.8974,
                        'apc_proximity_to_coding_v2': 27.1826,
                        'apc_proximity_to_tsstes': 9.32168,
                        'apc_transcription_factor': 13.4363,
                        'bravo_an': '264690',
                        'bravo_af': '0.530515',
                        'filter_status': 'PASS',
                        'polyphen2_hdiv_score': '0',
                        'polyphen2_hvar_score': '0',
                        'mutation_taster_score': '1',
                        'metasvm_pred': 'T',
                        'fathmm_xf': '1.3152',
                        'funseq_value': '1',
                        'funseq_description': 'coding'
                    },
                    'format': 'SNV;FAVORFullDB/genecode_comprehensive_exonic_info=PERM1:ENST00000341290.6:exon3:c.G1735C:p.E579Q,PERM1:ENST00000433179.3:exon1:c.G2077C:p.E693Q,;FAVORFullDB/genehancer=Name=1.11;genehancer_id=GH01I000975;connected_gene=GC01M000971;score=550.77;connected_gene=PERM1;score=14.82;connected_gene=ENSG00000272512;score=14.43;connected_gene=ISG15;score=11.60;connected_gene=LINC01786;score=11.48;connected_gene=ENSG00000240731;score=9.82;connected_gene=HES4;score=5.11;connected_gene=CDK11A;score=5.00;connected_gene=PLEKHN1;score=0.39;FAVORFullDB/af_total=0.523442;FAVORFullDB/af_asj_female=0.574915;FAVORFullDB/af_eas_female=0.825035;FAVORFullDB/af_afr_male=0.326112;FAVORFullDB/af_female=0.520831;FAVORFullDB/af_fin_male=0.560282;FAVORFullDB/af_oth_female=0.540984;FAVORFullDB/af_ami=0.424276;FAVORFullDB/af_oth=0.547398;FAVORFullDB/af_male=0.52622;FAVORFullDB/af_ami_female=0.410638;FAVORFullDB/af_afr=0.331292;FAVORFullDB/af_eas_male=0.812128;FAVORFullDB/af_sas=0.620814;FAVORFullDB/af_nfe_female=0.596343;FAVORFullDB/af_asj_male=0.594231;FAVORFullDB/af_raw=0.523451;FAVORFullDB/af_oth_male=0.55408;FAVORFullDB/af_nfe_male=0.588615;FAVORFullDB/af_asj=0.583986;FAVORFullDB/af_amr_male=0.664645;FAVORFullDB/af_amr_female=0.643147;FAVORFullDB/af_sas_female=0.602837;FAVORFullDB/af_fin=0.560693;FAVORFullDB/af_afr_female=0.3357;FAVORFullDB/af_sas_male=0.624899;FAVORFullDB/af_amr=0.655344;FAVORFullDB/af_nfe=0.593091;FAVORFullDB/af_eas=0.818095;FAVORFullDB/af_ami_male=0.439252;FAVORFullDB/af_fin_female=0.562;FAVORFullDB/linsight=0.214926;FAVORFullDB/gc=0.735;FAVORFullDB/cpg=0.133;FAVORFullDB/min_dist_tss=2077;FAVORFullDB/min_dist_tse=3748;FAVORFullDB/sift_cat=tolerated;FAVORFullDB/sift_val=0.47;FAVORFullDB/polyphen_cat=benign;FAVORFullDB/polyphen_val=0;FAVORFullDB/priphcons=0.013;FAVORFullDB/mamphcons=0;FAVORFullDB/verphcons=0;FAVORFullDB/priphylop=0.588;FAVORFullDB/mamphylop=-0.259;FAVORFullDB/verphylop=0.085;FAVORFullDB/bstatistic=866;FAVORFullDB/chmm_e1=1;FAVORFullDB/chmm_e2=0;FAVORFullDB/chmm_e3=0;FAVORFullDB/chmm_e4=1;FAVORFullDB/chmm_e5=1;FAVORFullDB/chmm_e6=0;FAVORFullDB/chmm_e7=0;FAVORFullDB/chmm_e8=0;FAVORFullDB/chmm_e9=0;FAVORFullDB/chmm_e10=0;FAVORFullDB/chmm_e11=0;FAVORFullDB/chmm_e12=0;FAVORFullDB/chmm_e13=0;FAVORFullDB/chmm_e14=0;FAVORFullDB/chmm_e15=12;FAVORFullDB/chmm_e16=1;FAVORFullDB/chmm_e17=1;FAVORFullDB/chmm_e18=0;FAVORFullDB/chmm_e19=1;FAVORFullDB/chmm_e20=1;FAVORFullDB/chmm_e21=16;FAVORFullDB/chmm_e22=4;FAVORFullDB/chmm_e23=7;FAVORFullDB/chmm_e24=2;FAVORFullDB/chmm_e25=0;FAVORFullDB/gerp_n=7.55;FAVORFullDB/gerp_s=-15.1;FAVORFullDB/encodeh3k4me1_sum=3.15;FAVORFullDB/encodeh3k4me2_sum=3.83;FAVORFullDB/encodeh3k4me3_sum=6.34;FAVORFullDB/encodeh3k9ac_sum=8.84;FAVORFullDB/encodeh3k9me3_sum=8.78;FAVORFullDB/encodeh3k27ac_sum=2.98;FAVORFullDB/encodeh3k27me3_sum=8.2;FAVORFullDB/encodeh3k36me3_sum=1.43;FAVORFullDB/encodeh3k79me2_sum=1.93;FAVORFullDB/encodeh4k20me1_sum=6.57;FAVORFullDB/encodeh2afz_sum=5.34;FAVORFullDB/encode_dnase_sum=0.57;FAVORFullDB/encodetotal_rna_sum=0.03;FAVORFullDB/grantham=29;FAVORFullDB/freq100bp=1;FAVORFullDB/rare100bp=4;FAVORFullDB/sngl100bp=20;FAVORFullDB/freq1000bp=3;FAVORFullDB/rare1000bp=15;FAVORFullDB/sngl1000bp=232;FAVORFullDB/freq10000bp=33;FAVORFullDB/rare10000bp=98;FAVORFullDB/sngl10000bp=1795;FAVORFullDB/remap_overlap_tf=20;FAVORFullDB/remap_overlap_cl=21;FAVORFullDB/cadd_rawscore=0.139414;FAVORFullDB/cadd_phred=3.277;FAVORFullDB/k24_bismap=0.292;FAVORFullDB/k24_umap=1;FAVORFullDB/k36_bismap=1;FAVORFullDB/k36_umap=1;FAVORFullDB/k50_bismap=1;FAVORFullDB/k50_umap=1;FAVORFullDB/k100_bismap=1;FAVORFullDB/k100_umap=1;FAVORFullDB/nucdiv=5.3;FAVORFullDB/rdhs=EH38D2115542;FAVORFullDB/recombination_rate=2.059;FAVORFullDB/refseq_category=exonic;FAVORFullDB/refseq_info=PERM1;FAVORFullDB/refseq_exonic_category=nonsynonymous',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 1.36453e-13,
                'slope': -0.735844,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'f50d42c0bf357fe71fbd0f5ddcd73efbcd48efe10ae2189e28a578d58036aaff',
                    '_id': 'variants/f50d42c0bf357fe71fbd0f5ddcd73efbcd48efe10ae2189e28a578d58036aaff',
                    '_rev': '_geyXI7---Z',
                    'chr': 'chr1',
                    'pos:long': 979471,
                    'rsid': [
                        'rs13303368'
                    ],
                    'ref': 'G',
                    'alt': 'C',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4874,
                                'alt:long': 0.5126
                            },
                            'alspac': {
                                'ref:long': 0.3871,
                                'alt:long': 0.6129
                            },
                            'estonian': {
                                'ref:long': 0.4616,
                                'alt:long': 0.5384
                            },
                            'exac': {
                                'ref:long': 0.405,
                                'alt:long': 0.595
                            },
                            'genome_dk': {
                                'ref:long': 0.45,
                                'alt:long': 0.55
                            },
                            'korean': {
                                'ref:long': 0.2551,
                                'alt:long': 0.7449
                            },
                            'korea1k': {
                                'ref:long': 0.2423,
                                'alt:long': 0.7577
                            },
                            'northernsweden': {
                                'ref:long': 0.42,
                                'alt:long': 0.58
                            },
                            'prjeb36033': {
                                'ref:long': 0.4,
                                'alt:long': 0.6
                            },
                            'qatari': {
                                'ref:long': 0.6574,
                                'alt:long': 0.3426
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2621,
                                'alt:long': 0.7379
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0.7273
                            },
                            'tommo': {
                                'ref:long': 0.249,
                                'alt:long': 0.751
                            },
                            'topmed': {
                                'ref:long': 0.5019,
                                'alt:long': 0.4981
                            },
                            'twinsuk': {
                                'ref:long': 0.3894,
                                'alt:long': 0.6106
                            },
                            'vietnamese': {
                                'ref:long': 0.2714,
                                'alt:long': 0.7286
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4361,
                                'alt:long': 0.5639
                            }
                        },
                        'varinfo': '1-979472-G-C',
                        'vid': '2.45842e+06',
                        'variant_vcf': '1-979472-G-C',
                        'variant_annovar': '1-979472-979472-G-C',
                        'start_position': '979472',
                        'end_position': '979472',
                        'ref_annovar': 'G',
                        'alt_annovar': 'C',
                        'ref_vcf': 'G',
                        'alt_vcf': 'C',
                        'aloft_value': '0.4247',
                        'aloft_description': 'Tolerant',
                        'apc_conservation': 2.9648,
                        'apc_conservation_v2': 6.34156,
                        'apc_epigenetics_active': 1.84411,
                        'apc_epigenetics': 3.06528,
                        'apc_epigenetics_repressed': 0.736671,
                        'apc_epigenetics_transcription': 0.881132,
                        'apc_local_nucleotide_diversity': 2.70482,
                        'apc_local_nucleotide_diversity_v2': 0.0299207,
                        'apc_local_nucleotide_diversity_v3': 0.0378455,
                        'apc_mappability': 9.95131,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0297314,
                        'apc_protein_function': 20.7294,
                        'apc_protein_function_v2': 20.7292,
                        'apc_protein_function_v3': 20.7293,
                        'apc_proximity_to_coding': 23.5039,
                        'apc_proximity_to_coding_v2': 23.6763,
                        'apc_proximity_to_tsstes': 9.59122,
                        'apc_transcription_factor': 11.3632,
                        'bravo_an': '264690',
                        'bravo_af': '0.498111',
                        'filter_status': 'PASS',
                        'polyphen2_hdiv_score': '0',
                        'polyphen2_hvar_score': '0',
                        'mutation_taster_score': '1',
                        'metasvm_pred': 'T',
                        'fathmm_xf': '2.13635',
                        'funseq_value': '1',
                        'funseq_description': 'coding'
                    },
                    'format': 'SNV;FAVORFullDB/genecode_comprehensive_exonic_info=PERM1:ENST00000341290.6:exon3:c.C1216G:p.Q406E,PERM1:ENST00000433179.3:exon1:c.C1558G:p.Q520E,;FAVORFullDB/genehancer=Name=1.11;genehancer_id=GH01I000975;connected_gene=GC01M000971;score=550.77;connected_gene=PERM1;score=14.82;connected_gene=ENSG00000272512;score=14.43;connected_gene=ISG15;score=11.60;connected_gene=LINC01786;score=11.48;connected_gene=ENSG00000240731;score=9.82;connected_gene=HES4;score=5.11;connected_gene=CDK11A;score=5.00;connected_gene=PLEKHN1;score=0.39;FAVORFullDB/af_total=0.493309;FAVORFullDB/af_asj_female=0.568104;FAVORFullDB/af_eas_female=0.718923;FAVORFullDB/af_afr_male=0.261484;FAVORFullDB/af_female=0.491272;FAVORFullDB/af_fin_male=0.520881;FAVORFullDB/af_oth_female=0.509107;FAVORFullDB/af_ami=0.423333;FAVORFullDB/af_oth=0.512093;FAVORFullDB/af_male=0.495476;FAVORFullDB/af_ami_female=0.410638;FAVORFullDB/af_afr=0.265059;FAVORFullDB/af_eas_male=0.728029;FAVORFullDB/af_sas=0.612935;FAVORFullDB/af_nfe_female=0.591038;FAVORFullDB/af_asj_male=0.583867;FAVORFullDB/af_raw=0.493409;FAVORFullDB/af_oth_male=0.515209;FAVORFullDB/af_nfe_male=0.584026;FAVORFullDB/af_asj=0.575511;FAVORFullDB/af_amr_male=0.635976;FAVORFullDB/af_amr_female=0.613836;FAVORFullDB/af_sas_female=0.593972;FAVORFullDB/af_fin=0.523636;FAVORFullDB/af_afr_female=0.268108;FAVORFullDB/af_sas_male=0.617244;FAVORFullDB/af_amr=0.626391;FAVORFullDB/af_nfe=0.588086;FAVORFullDB/af_eas=0.723819;FAVORFullDB/af_ami_male=0.437209;FAVORFullDB/af_fin_female=0.5324;FAVORFullDB/linsight=0.214926;FAVORFullDB/gc=0.695;FAVORFullDB/cpg=0.067;FAVORFullDB/min_dist_tss=1558;FAVORFullDB/min_dist_tse=4267;FAVORFullDB/sift_cat=tolerated;FAVORFullDB/sift_val=1;FAVORFullDB/polyphen_cat=benign;FAVORFullDB/polyphen_val=0;FAVORFullDB/priphcons=0;FAVORFullDB/mamphcons=0;FAVORFullDB/verphcons=0;FAVORFullDB/priphylop=0.413;FAVORFullDB/mamphylop=0.194;FAVORFullDB/verphylop=0.291;FAVORFullDB/bstatistic=866;FAVORFullDB/chmm_e1=0;FAVORFullDB/chmm_e2=1;FAVORFullDB/chmm_e3=0;FAVORFullDB/chmm_e4=1;FAVORFullDB/chmm_e5=1;FAVORFullDB/chmm_e6=0;FAVORFullDB/chmm_e7=0;FAVORFullDB/chmm_e8=0;FAVORFullDB/chmm_e9=0;FAVORFullDB/chmm_e10=0;FAVORFullDB/chmm_e11=0;FAVORFullDB/chmm_e12=0;FAVORFullDB/chmm_e13=0;FAVORFullDB/chmm_e14=0;FAVORFullDB/chmm_e15=13;FAVORFullDB/chmm_e16=0;FAVORFullDB/chmm_e17=0;FAVORFullDB/chmm_e18=0;FAVORFullDB/chmm_e19=2;FAVORFullDB/chmm_e20=0;FAVORFullDB/chmm_e21=17;FAVORFullDB/chmm_e22=4;FAVORFullDB/chmm_e23=7;FAVORFullDB/chmm_e24=1;FAVORFullDB/chmm_e25=1;FAVORFullDB/gerp_n=12;FAVORFullDB/gerp_s=-24;FAVORFullDB/encodeh3k4me1_sum=1.89;FAVORFullDB/encodeh3k4me2_sum=2.01;FAVORFullDB/encodeh3k4me3_sum=3.82;FAVORFullDB/encodeh3k9ac_sum=3.93;FAVORFullDB/encodeh3k9me3_sum=1.56;FAVORFullDB/encodeh3k27ac_sum=1.96;FAVORFullDB/encodeh3k27me3_sum=9.72;FAVORFullDB/encodeh3k36me3_sum=1.52;FAVORFullDB/encodeh3k79me2_sum=2.41;FAVORFullDB/encodeh4k20me1_sum=5.67;FAVORFullDB/encodeh2afz_sum=4.73;FAVORFullDB/encode_dnase_sum=0.52;FAVORFullDB/grantham=29;FAVORFullDB/freq100bp=1;FAVORFullDB/rare100bp=5;FAVORFullDB/sngl100bp=28;FAVORFullDB/freq1000bp=3;FAVORFullDB/rare1000bp=14;FAVORFullDB/sngl1000bp=235;FAVORFullDB/freq10000bp=34;FAVORFullDB/rare10000bp=96;FAVORFullDB/sngl10000bp=1788;FAVORFullDB/remap_overlap_tf=12;FAVORFullDB/remap_overlap_cl=13;FAVORFullDB/cadd_rawscore=-0.226183;FAVORFullDB/cadd_phred=0.45;FAVORFullDB/k24_bismap=0.75;FAVORFullDB/k24_umap=1;FAVORFullDB/k36_bismap=1;FAVORFullDB/k36_umap=1;FAVORFullDB/k50_bismap=1;FAVORFullDB/k50_umap=1;FAVORFullDB/k100_bismap=1;FAVORFullDB/k100_umap=1;FAVORFullDB/nucdiv=5.4;FAVORFullDB/recombination_rate=2.059;FAVORFullDB/refseq_category=exonic;FAVORFullDB/refseq_info=PERM1;FAVORFullDB/refseq_exonic_category=nonsynonymous',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 7.81183e-12,
                'slope': -0.68954,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '362436e9df3d27ccbbb9713cde1fc5e0e896446113822a011a8804bfe3a35f7d',
                    '_id': 'variants/362436e9df3d27ccbbb9713cde1fc5e0e896446113822a011a8804bfe3a35f7d',
                    '_rev': '_geyXI66--G',
                    'chr': 'chr1',
                    'pos:long': 979559,
                    'rsid': [
                        'rs13303033'
                    ],
                    'ref': 'T',
                    'alt': 'C',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4904,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.4014,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.4714,
                                'alt:long': 0
                            },
                            'exac': {
                                'ref:long': 0.4126,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.475,
                                'alt:long': 0
                            },
                            'gnomad': {
                                'ref:long': 0.5185,
                                'alt:long': 0
                            },
                            'gnomad_exomes': {
                                'ref:long': 0.4113,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.4078,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.2538,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.2429,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.43,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.6528,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2654,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.2493,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.5107,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.4048,
                                'alt:long': 0
                            },
                            'vietnamese': {
                                'ref:long': 0.2917,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4484,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-979560-T-C',
                        'vid': '2.45868e+06',
                        'variant_vcf': '1-979560-T-C',
                        'variant_annovar': '1-979560-979560-T-C',
                        'start_position': '979560',
                        'end_position': '979560',
                        'ref_annovar': 'T',
                        'alt_annovar': 'C',
                        'ref_vcf': 'T',
                        'alt_vcf': 'C',
                        'apc_conservation': 0.72745,
                        'apc_conservation_v2': 2.73536,
                        'apc_epigenetics_active': 1.38995,
                        'apc_epigenetics': 3.8763,
                        'apc_epigenetics_repressed': 1.36417,
                        'apc_epigenetics_transcription': 5.2601,
                        'apc_local_nucleotide_diversity': 2.79722,
                        'apc_local_nucleotide_diversity_v2': 0.0978268,
                        'apc_local_nucleotide_diversity_v3': 0.107916,
                        'apc_mappability': 12.6226,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0973021,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 23.103,
                        'apc_proximity_to_coding_v2': 23.2603,
                        'apc_proximity_to_tsstes': 9.65537,
                        'apc_transcription_factor': 11.6783,
                        'bravo_an': '264690',
                        'bravo_af': '0.489323',
                        'filter_status': 'PASS',
                        'fathmm_xf': '0.405956',
                        'funseq_value': '0',
                        'funseq_description': 'coding'
                    },
                    'format': 'SNV;FAVORFullDB/genecode_comprehensive_exonic_info=PERM1:ENST00000341290.6:exon3:c.A1128G:p.A376A,PERM1:ENST00000433179.3:exon1:c.A1470G:p.A490A,;FAVORFullDB/genehancer=Name=1.11;genehancer_id=GH01I000975;connected_gene=GC01M000971;score=550.77;connected_gene=PERM1;score=14.82;connected_gene=ENSG00000272512;score=14.43;connected_gene=ISG15;score=11.60;connected_gene=LINC01786;score=11.48;connected_gene=ENSG00000240731;score=9.82;connected_gene=HES4;score=5.11;connected_gene=CDK11A;score=5.00;connected_gene=PLEKHN1;score=0.39;FAVORFullDB/af_total=0.484311;FAVORFullDB/af_asj_female=0.567537;FAVORFullDB/af_eas_female=0.719225;FAVORFullDB/af_afr_male=0.256655;FAVORFullDB/af_female=0.481073;FAVORFullDB/af_fin_male=0.520534;FAVORFullDB/af_oth_female=0.503643;FAVORFullDB/af_ami=0.423333;FAVORFullDB/af_oth=0.506977;FAVORFullDB/af_male=0.487758;FAVORFullDB/af_ami_female=0.410638;FAVORFullDB/af_afr=0.260095;FAVORFullDB/af_eas_male=0.728944;FAVORFullDB/af_sas=0.613741;FAVORFullDB/af_nfe_female=0.574922;FAVORFullDB/af_asj_male=0.583867;FAVORFullDB/af_raw=0.48427;FAVORFullDB/af_oth_male=0.510456;FAVORFullDB/af_nfe_male=0.569171;FAVORFullDB/af_asj=0.575211;FAVORFullDB/af_amr_male=0.631002;FAVORFullDB/af_amr_female=0.609851;FAVORFullDB/af_sas_female=0.593972;FAVORFullDB/af_fin=0.523285;FAVORFullDB/af_afr_female=0.263023;FAVORFullDB/af_sas_male=0.618241;FAVORFullDB/af_amr=0.621851;FAVORFullDB/af_nfe=0.572502;FAVORFullDB/af_eas=0.724457;FAVORFullDB/af_ami_male=0.437209;FAVORFullDB/af_fin_female=0.532026;FAVORFullDB/linsight=0.214926;FAVORFullDB/gc=0.649;FAVORFullDB/cpg=0.067;FAVORFullDB/min_dist_tss=1470;FAVORFullDB/min_dist_tse=4355;FAVORFullDB/priphcons=0.003;FAVORFullDB/mamphcons=0;FAVORFullDB/verphcons=0;FAVORFullDB/priphylop=0.402;FAVORFullDB/mamphylop=-0.569;FAVORFullDB/verphylop=-1.136;FAVORFullDB/bstatistic=861;FAVORFullDB/chmm_e1=0;FAVORFullDB/chmm_e2=1;FAVORFullDB/chmm_e3=0;FAVORFullDB/chmm_e4=1;FAVORFullDB/chmm_e5=1;FAVORFullDB/chmm_e6=0;FAVORFullDB/chmm_e7=0;FAVORFullDB/chmm_e8=0;FAVORFullDB/chmm_e9=0;FAVORFullDB/chmm_e10=0;FAVORFullDB/chmm_e11=0;FAVORFullDB/chmm_e12=0;FAVORFullDB/chmm_e13=0;FAVORFullDB/chmm_e14=0;FAVORFullDB/chmm_e15=13;FAVORFullDB/chmm_e16=0;FAVORFullDB/chmm_e17=0;FAVORFullDB/chmm_e18=0;FAVORFullDB/chmm_e19=2;FAVORFullDB/chmm_e20=0;FAVORFullDB/chmm_e21=17;FAVORFullDB/chmm_e22=4;FAVORFullDB/chmm_e23=7;FAVORFullDB/chmm_e24=1;FAVORFullDB/chmm_e25=1;FAVORFullDB/gerp_n=13.2;FAVORFullDB/gerp_s=-26.4;FAVORFullDB/encodeh3k4me1_sum=2.5;FAVORFullDB/encodeh3k4me2_sum=0.31;FAVORFullDB/encodeh3k4me3_sum=2.52;FAVORFullDB/encodeh3k9ac_sum=4.15;FAVORFullDB/encodeh3k9me3_sum=2.35;FAVORFullDB/encodeh3k27ac_sum=2.88;FAVORFullDB/encodeh3k27me3_sum=11.97;FAVORFullDB/encodeh3k36me3_sum=5.81;FAVORFullDB/encodeh3k79me2_sum=1.99;FAVORFullDB/encodeh4k20me1_sum=9.5;FAVORFullDB/encodeh2afz_sum=3.98;FAVORFullDB/encode_dnase_sum=0.75;FAVORFullDB/encodetotal_rna_sum=0;FAVORFullDB/freq100bp=1;FAVORFullDB/rare100bp=0;FAVORFullDB/sngl100bp=19;FAVORFullDB/freq1000bp=3;FAVORFullDB/rare1000bp=14;FAVORFullDB/sngl1000bp=229;FAVORFullDB/freq10000bp=35;FAVORFullDB/rare10000bp=95;FAVORFullDB/sngl10000bp=1791;FAVORFullDB/remap_overlap_tf=13;FAVORFullDB/remap_overlap_cl=14;FAVORFullDB/cadd_rawscore=-0.032936;FAVORFullDB/cadd_phred=1.471;FAVORFullDB/k24_bismap=0.833;FAVORFullDB/k24_umap=1;FAVORFullDB/k36_bismap=1;FAVORFullDB/k36_umap=1;FAVORFullDB/k50_bismap=1;FAVORFullDB/k50_umap=1;FAVORFullDB/k100_bismap=1;FAVORFullDB/k100_umap=1;FAVORFullDB/nucdiv=5.45;FAVORFullDB/recombination_rate=2.059;FAVORFullDB/refseq_category=exonic;FAVORFullDB/refseq_info=PERM1;FAVORFullDB/refseq_exonic_category=synonymous',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 4.87218e-14,
                'slope': -0.743714,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '22f170e54c30a59e737beba20444f192201126f0b1415a7c9a106d1d01fe98d0',
                    '_id': 'variants/22f170e54c30a59e737beba20444f192201126f0b1415a7c9a106d1d01fe98d0',
                    '_rev': '_geyXJ-S--K',
                    'chr': 'chr1',
                    'pos:long': 981168,
                    'rsid': [
                        'rs6660139'
                    ],
                    'ref': 'A',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.2414,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.2385,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.2804,
                                'alt:long': 0
                            },
                            'exac': {
                                'ref:long': 0.2164,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.225,
                                'alt:long': 0
                            },
                            'gnomad': {
                                'ref:long': 0.2855,
                                'alt:long': 0
                            },
                            'gnomad_exomes': {
                                'ref:long': 0.2265,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.2355,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.1307,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.1119,
                                'alt:long': 0
                            },
                            'mgp': {
                                'ref:long': 0.2341,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.28,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.2778,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.15,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.1538,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.09564,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.28,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.2543,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.2683,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-981169-A-G',
                        'vid': '2.46351e+06',
                        'variant_vcf': '1-981169-A-G',
                        'variant_annovar': '1-981169-981169-A-G',
                        'start_position': '981169',
                        'end_position': '981169',
                        'ref_annovar': 'A',
                        'alt_annovar': 'G',
                        'ref_vcf': 'A',
                        'alt_vcf': 'G',
                        'apc_conservation': 0.54059,
                        'apc_conservation_v2': 0.987718,
                        'apc_epigenetics_active': 9.29603,
                        'apc_epigenetics': 10.9147,
                        'apc_epigenetics_repressed': 5.50979,
                        'apc_epigenetics_transcription': 3.20664,
                        'apc_local_nucleotide_diversity': 3.00724,
                        'apc_local_nucleotide_diversity_v2': 0.0719121,
                        'apc_local_nucleotide_diversity_v3': 0.0845642,
                        'apc_mappability': 6.72026,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0713297,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 14.076,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 13.3714,
                        'apc_transcription_factor': 9.69058,
                        'bravo_an': '264690',
                        'bravo_af': '0.720035',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.18201',
                        'funseq_value': '1',
                        'funseq_description': 'coding',
                        'af_total': 0.716829,
                        'af_asj_female': 0.733825,
                        'af_eas_female': 0.881579,
                        'af_afr_male': 0.621451,
                        'af_female': 0.714635,
                        'af_fin_male': 0.730779,
                        'af_oth_female': 0.739927,
                        'af_ami': 0.707589,
                        'af_oth': 0.734608,
                        'af_male': 0.719163,
                        'af_ami_female': 0.693617,
                        'af_afr': 0.623712,
                        'af_eas_male': 0.871429,
                        'af_sas': 0.825972,
                        'af_nfe_female': 0.744993,
                        'af_asj_male': 0.726633,
                        'af_raw': 0.716789,
                        'af_oth_male': 0.729087,
                        'af_nfe_male': 0.743827,
                        'af_asj': 0.730445,
                        'af_amr_male': 0.794057,
                        'af_amr_female': 0.790437,
                        'af_fin': 0.734477,
                        'af_afr_female': 0.625641,
                        'af_sas_male': 0.826456,
                        'af_amr': 0.792492,
                        'af_nfe': 0.744502,
                        'af_eas': 0.87612,
                        'af_ami_male': 0.723005,
                        'af_fin_female': 0.746203,
                        'cadd_rawscore': '-0.016826',
                        'cadd_phred': '1.604',
                        'refseq_category': 'UTR5',
                        'tg_afr': '0.5998',
                        'tg_all': '0.758586',
                        'tg_amr': '0.781',
                        'tg_eas': '0.88',
                        'tg_eur': '0.7584',
                        'tg_sas': '0.8323'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 2.06171e-11,
                'slope': -0.814103,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '0e82ef1610de3c07a4b9a57ac0ccf090da31cd1bba048a25b5680c41f950dbf5',
                    '_id': 'variants/0e82ef1610de3c07a4b9a57ac0ccf090da31cd1bba048a25b5680c41f950dbf5',
                    '_rev': '_geyXJ-W--c',
                    'chr': 'chr1',
                    'pos:long': 981453,
                    'rsid': [
                        'rs6694632'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4469,
                                'alt:long': 0.5531
                            },
                            'alspac': {
                                'ref:long': 0.3996,
                                'alt:long': 0.6004
                            },
                            'estonian': {
                                'ref:long': 0.454,
                                'alt:long': 0.546
                            },
                            'genome_dk': {
                                'ref:long': 0.475,
                                'alt:long': 0.525
                            },
                            'gnomad': {
                                'ref:long': 0.4849,
                                'alt:long': 0.5151
                            },
                            'gonl': {
                                'ref:long': 0.4018,
                                'alt:long': 0.5982
                            },
                            'hapmap': {
                                'ref:long': 0.4847,
                                'alt:long': 0.5153
                            },
                            'korean': {
                                'ref:long': 0.1758,
                                'alt:long': 0.8242
                            },
                            'korea1k': {
                                'ref:long': 0.1681,
                                'alt:long': 0.8319
                            },
                            'northernsweden': {
                                'ref:long': 0.41,
                                'alt:long': 0.59
                            },
                            'qatari': {
                                'ref:long': 0.6343,
                                'alt:long': 0.3657
                            },
                            'sgdp_prj': {
                                'ref:long': 0.25,
                                'alt:long': 0.75
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0.7273
                            },
                            'tommo': {
                                'ref:long': 0.1744,
                                'alt:long': 0.8256
                            },
                            'topmed': {
                                'ref:long': 0.4759,
                                'alt:long': 0.5241
                            },
                            'twinsuk': {
                                'ref:long': 0.401,
                                'alt:long': 0.599
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4473,
                                'alt:long': 0.5527
                            }
                        },
                        'varinfo': '1-981454-G-A',
                        'vid': '2.46436e+06',
                        'variant_vcf': '1-981454-G-A',
                        'variant_annovar': '1-981454-981454-G-A',
                        'start_position': '981454',
                        'end_position': '981454',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 2.14319,
                        'apc_conservation_v2': 4.85704,
                        'apc_epigenetics_active': 12.5908,
                        'apc_epigenetics': 13.0979,
                        'apc_epigenetics_repressed': 8.51195,
                        'apc_epigenetics_transcription': 2.78626,
                        'apc_local_nucleotide_diversity': 3.00724,
                        'apc_local_nucleotide_diversity_v2': 0.10387,
                        'apc_local_nucleotide_diversity_v3': 0.11657,
                        'apc_mappability': 9.95131,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.103199,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 8.71688,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 11.1589,
                        'apc_transcription_factor': 11.2487,
                        'bravo_an': '264690',
                        'bravo_af': '0.524077',
                        'filter_status': 'PASS',
                        'fathmm_xf': '0.416403',
                        'funseq_description': 'noncoding',
                        'af_total': 0.517371,
                        'af_asj_female': 0.572645,
                        'af_eas_female': 0.824343,
                        'af_afr_male': 0.328807,
                        'af_female': 0.513369,
                        'af_fin_male': 0.558609,
                        'af_oth_female': 0.538251,
                        'af_ami': 0.423163,
                        'af_oth': 0.542326,
                        'af_male': 0.521629,
                        'af_ami_female': 0.410256,
                        'af_afr': 0.332817,
                        'af_eas_male': 0.81369,
                        'af_sas': 0.622536,
                        'af_nfe_female': 0.581522,
                        'af_asj_male': 0.594994,
                        'af_raw': 0.517427,
                        'af_oth_male': 0.546578,
                        'af_nfe_male': 0.57573,
                        'af_asj': 0.583133,
                        'af_amr_male': 0.664256,
                        'af_amr_female': 0.641794,
                        'af_fin': 0.559291,
                        'af_afr_female': 0.336234,
                        'af_sas_male': 0.626312,
                        'af_amr': 0.654533,
                        'af_nfe': 0.579084,
                        'af_eas': 0.818618,
                        'af_ami_male': 0.437209,
                        'af_fin_female': 0.561453,
                        'cadd_rawscore': '-0.011885',
                        'cadd_phred': '1.646',
                        'refseq_category': 'intronic',
                        'tg_afr': '0.239',
                        'tg_all': '0.553115',
                        'tg_amr': '0.6499',
                        'tg_eas': '0.8056',
                        'tg_eur': '0.5954',
                        'tg_sas': '0.6053'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 1.18979e-15,
                'slope': -0.781824,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '85056aaee88e687530ffafe66566942b2ea4c97cc1e9ad6b8d3316e0c2ca03d7',
                    '_id': 'variants/85056aaee88e687530ffafe66566942b2ea4c97cc1e9ad6b8d3316e0c2ca03d7',
                    '_rev': '_geyXJ-y--W',
                    'chr': 'chr1',
                    'pos:long': 982259,
                    'rsid': [
                        'rs41285816'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8486,
                                'alt:long': 0.1514
                            },
                            'alspac': {
                                'ref:long': 0.788,
                                'alt:long': 0.212
                            },
                            'estonian': {
                                'ref:long': 0.7458,
                                'alt:long': 0.2542
                            },
                            'genome_dk': {
                                'ref:long': 0.8,
                                'alt:long': 0.2
                            },
                            'gnomad': {
                                'ref:long': 0.8105,
                                'alt:long': 0.1895
                            },
                            'gonl': {
                                'ref:long': 0.7786,
                                'alt:long': 0.2214
                            },
                            'korean': {
                                'ref:long': 0.8719,
                                'alt:long': 0.1281
                            },
                            'korea1k': {
                                'ref:long': 0.8886,
                                'alt:long': 0.1114
                            },
                            'northernsweden': {
                                'ref:long': 0.7417,
                                'alt:long': 0.2583
                            },
                            'page_study': {
                                'ref:long': 0.8586,
                                'alt:long': 0.1414
                            },
                            'qatari': {
                                'ref:long': 0.838,
                                'alt:long': 0.162
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4247,
                                'alt:long': 0.5753
                            },
                            'siberian': {
                                'ref:long': 0.375,
                                'alt:long': 0.625
                            },
                            'tommo': {
                                'ref:long': 0.9051,
                                'alt:long': 0.09494
                            },
                            'topmed': {
                                'ref:long': 0.8206,
                                'alt:long': 0.1794
                            },
                            'twinsuk': {
                                'ref:long': 0.778,
                                'alt:long': 0.222
                            },
                            'vietnamese': {
                                'ref:long': 0.912,
                                'alt:long': 0.08796
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.7952,
                                'alt:long': 0.2048
                            }
                        },
                        'varinfo': '1-982260-G-A',
                        'vid': '2.46678e+06',
                        'variant_vcf': '1-982260-G-A',
                        'variant_annovar': '1-982260-982260-G-A',
                        'start_position': '982260',
                        'end_position': '982260',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 7.54392,
                        'apc_conservation_v2': 7.70236,
                        'apc_epigenetics_active': 13.8934,
                        'apc_epigenetics': 15.1616,
                        'apc_epigenetics_repressed': 7.98868,
                        'apc_epigenetics_transcription': 4.26696,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.108777,
                        'apc_local_nucleotide_diversity_v3': 0.124837,
                        'apc_mappability': 17.0936,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.107814,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 12.2324,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 12.6511,
                        'apc_transcription_factor': 13.3996,
                        'bravo_an': '264690',
                        'bravo_af': '0.17941',
                        'filter_status': 'PASS',
                        'fathmm_xf': '11.6379',
                        'funseq_description': 'noncoding',
                        'af_total': 0.188816,
                        'af_asj_female': 0.18672,
                        'af_eas_female': 0.119475,
                        'af_afr_male': 0.134106,
                        'af_female': 0.188287,
                        'af_fin_male': 0.261587,
                        'af_oth_female': 0.176364,
                        'af_ami': 0.268374,
                        'af_oth': 0.18013,
                        'af_male': 0.189379,
                        'af_ami_female': 0.275641,
                        'af_afr': 0.135719,
                        'af_eas_male': 0.128707,
                        'af_sas': 0.157912,
                        'af_nfe_female': 0.223867,
                        'af_asj_male': 0.201665,
                        'af_raw': 0.188993,
                        'af_oth_male': 0.184061,
                        'af_nfe_male': 0.223404,
                        'af_asj': 0.193742,
                        'af_amr_male': 0.151989,
                        'af_amr_female': 0.152049,
                        'af_fin': 0.256943,
                        'af_afr_female': 0.137093,
                        'af_sas_male': 0.156452,
                        'af_amr': 0.152015,
                        'af_nfe': 0.223672,
                        'af_eas': 0.124442,
                        'af_ami_male': 0.260465,
                        'af_fin_female': 0.242206,
                        'cadd_rawscore': '0.346334',
                        'cadd_phred': '5.819',
                        'refseq_category': 'upstream',
                        'tg_afr': '0.1346',
                        'tg_all': '0.151358',
                        'tg_amr': '0.1571',
                        'tg_eas': '0.119',
                        'tg_eur': '0.2008',
                        'tg_sas': '0.1524'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 6.63382e-9,
                'slope': 0.724306,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '2c709a08569f57d8532d95ee7a70082114fa12491f984cb8f336cf90bc18323f',
                    '_id': 'variants/2c709a08569f57d8532d95ee7a70082114fa12491f984cb8f336cf90bc18323f',
                    '_rev': '_geyXJ_---R',
                    'chr': 'chr1',
                    'pos:long': 982857,
                    'rsid': [
                        'rs28552711'
                    ],
                    'ref': 'C',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8442,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.7875,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.7455,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.7786,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.8712,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.8896,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.7417,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.838,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4247,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.375,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.905,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.819,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.7786,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8768,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-982858-C-G',
                        'vid': '2.46857e+06',
                        'variant_vcf': '1-982858-C-G',
                        'variant_annovar': '1-982858-982858-C-G',
                        'start_position': '982858',
                        'end_position': '982858',
                        'ref_annovar': 'C',
                        'alt_annovar': 'G',
                        'ref_vcf': 'C',
                        'alt_vcf': 'G',
                        'apc_conservation': 0.0176052,
                        'apc_conservation_v2': 0.0197938,
                        'apc_epigenetics_active': 10.4046,
                        'apc_epigenetics': 12.4107,
                        'apc_epigenetics_repressed': 23.3074,
                        'apc_epigenetics_transcription': 1.55535,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.0735644,
                        'apc_local_nucleotide_diversity_v3': 0.0845033,
                        'apc_mappability': 8.95186,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.073133,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 1.36735,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 9.83349,
                        'apc_transcription_factor': 10.9549,
                        'bravo_an': '264690',
                        'bravo_af': '0.18105',
                        'filter_status': 'PASS',
                        'fathmm_xf': '4.30481',
                        'funseq_description': 'noncoding',
                        'af_total': 0.190503,
                        'af_asj_female': 0.18672,
                        'af_eas_female': 0.119475,
                        'af_afr_male': 0.138791,
                        'af_female': 0.189677,
                        'af_fin_male': 0.26111,
                        'af_oth_female': 0.178182,
                        'af_ami': 0.269488,
                        'af_oth': 0.182621,
                        'af_male': 0.191382,
                        'af_ami_female': 0.280851,
                        'af_afr': 0.13972,
                        'af_eas_male': 0.128266,
                        'af_sas': 0.159344,
                        'af_nfe_female': 0.223824,
                        'af_asj_male': 0.201665,
                        'af_raw': 0.190636,
                        'af_oth_male': 0.187262,
                        'af_nfe_male': 0.223432,
                        'af_asj': 0.193742,
                        'af_amr_male': 0.157623,
                        'af_amr_female': 0.155668,
                        'af_fin': 0.256638,
                        'af_afr_female': 0.140511,
                        'af_sas_male': 0.158213,
                        'af_amr': 0.156777,
                        'af_nfe': 0.223659,
                        'af_eas': 0.124202,
                        'af_ami_male': 0.257009,
                        'af_fin_female': 0.242412,
                        'cadd_rawscore': '-0.368952',
                        'cadd_phred': '0.153',
                        'refseq_category': 'upstream',
                        'tg_afr': '0.1505',
                        'tg_all': '0.155751',
                        'tg_amr': '0.1571',
                        'tg_eas': '0.119',
                        'tg_eur': '0.2008',
                        'tg_sas': '0.1534'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 6.63382e-9,
                'slope': 0.724306,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '2cb8de7458336a8afcacca1ecb09be6bde9530710664af37d1b50e6064100a50',
                    '_id': 'variants/2cb8de7458336a8afcacca1ecb09be6bde9530710664af37d1b50e6064100a50',
                    '_rev': '_geyXJDG--B',
                    'chr': 'chr1',
                    'pos:long': 983003,
                    'rsid': [
                        'rs13303118'
                    ],
                    'ref': 'G',
                    'alt': 'T',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.5074,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.4022,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.4701,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.475,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.4078,
                                'alt:long': 0
                            },
                            'hapmap': {
                                'ref:long': 0.5653,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.2532,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.2391,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.43,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.6667,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2581,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.2495,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.536,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.4056,
                                'alt:long': 0
                            },
                            'vietnamese': {
                                'ref:long': 0.2897,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4442,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-983004-G-T',
                        'vid': '2.46901e+06',
                        'variant_vcf': '1-983004-G-T',
                        'variant_annovar': '1-983004-983004-G-T',
                        'start_position': '983004',
                        'end_position': '983004',
                        'ref_annovar': 'G',
                        'alt_annovar': 'T',
                        'ref_vcf': 'G',
                        'alt_vcf': 'T',
                        'apc_conservation': 0.734502,
                        'apc_conservation_v2': 0.905613,
                        'apc_epigenetics_active': 7.88419,
                        'apc_epigenetics': 11.1019,
                        'apc_epigenetics_repressed': 10.3892,
                        'apc_epigenetics_transcription': 5.24549,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.0611395,
                        'apc_local_nucleotide_diversity_v3': 0.0721988,
                        'apc_mappability': 11.1486,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0606441,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 4.22027,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 9.51103,
                        'apc_transcription_factor': 13.2021,
                        'bravo_an': '264690',
                        'bravo_af': '0.464041',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.87579',
                        'funseq_description': 'noncoding',
                        'af_total': 0.460338,
                        'af_asj_female': 0.567537,
                        'af_eas_female': 0.71706,
                        'af_afr_male': 0.178276,
                        'af_female': 0.456219,
                        'af_fin_male': 0.519386,
                        'af_oth_female': 0.482696,
                        'af_ami': 0.423333,
                        'af_oth': 0.487454,
                        'af_male': 0.46472,
                        'af_ami_female': 0.410638,
                        'af_afr': 0.181948,
                        'af_eas_male': 0.729537,
                        'af_sas': 0.614248,
                        'af_nfe_female': 0.574347,
                        'af_asj_male': 0.583974,
                        'af_raw': 0.460343,
                        'af_oth_male': 0.49241,
                        'af_nfe_male': 0.568694,
                        'af_asj': 0.575256,
                        'af_amr_male': 0.625871,
                        'af_amr_female': 0.605968,
                        'af_fin': 0.522392,
                        'af_afr_female': 0.185076,
                        'af_sas_male': 0.617339,
                        'af_amr': 0.617268,
                        'af_nfe': 0.571969,
                        'af_eas': 0.723785,
                        'af_ami_male': 0.437209,
                        'af_fin_female': 0.531923,
                        'cadd_rawscore': '-0.104997',
                        'cadd_phred': '0.975',
                        'refseq_category': 'upstream',
                        'tg_afr': '0.1112',
                        'tg_all': '0.492612',
                        'tg_amr': '0.6052',
                        'tg_eas': '0.7133',
                        'tg_eur': '0.5895',
                        'tg_sas': '0.6012'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 8.47956e-16,
                'slope': -0.801873,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'a769b391bc07d5f76c3a65005a506404cc976a34ece96795aa7d51c7586cb5f7',
                    '_id': 'variants/a769b391bc07d5f76c3a65005a506404cc976a34ece96795aa7d51c7586cb5f7',
                    '_rev': '_geyXJDK--O',
                    'chr': 'chr1',
                    'pos:long': 983192,
                    'rsid': [
                        'rs2341354'
                    ],
                    'ref': 'A',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4509,
                                'alt:long': 0.5491
                            },
                            'alspac': {
                                'ref:long': 0.3991,
                                'alt:long': 0.6009
                            },
                            'estonian': {
                                'ref:long': 0.4536,
                                'alt:long': 0.5464
                            },
                            'genome_dk': {
                                'ref:long': 0.475,
                                'alt:long': 0.525
                            },
                            'gnomad': {
                                'ref:long': 0.4915,
                                'alt:long': 0.5085
                            },
                            'gonl': {
                                'ref:long': 0.3988,
                                'alt:long': 0.6012
                            },
                            'hapmap': {
                                'ref:long': 0.4909,
                                'alt:long': 0.5091
                            },
                            'korean': {
                                'ref:long': 0.1751,
                                'alt:long': 0.8249
                            },
                            'korea1k': {
                                'ref:long': 0.1643,
                                'alt:long': 0.8357
                            },
                            'northernsweden': {
                                'ref:long': 0.41,
                                'alt:long': 0.59
                            },
                            'page_study': {
                                'ref:long': 0.4832,
                                'alt:long': 0.5168
                            },
                            'prjeb37584': {
                                'ref:long': 0.1816,
                                'alt:long': 0.8184
                            },
                            'qatari': {
                                'ref:long': 0.6389,
                                'alt:long': 0.3611
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2489,
                                'alt:long': 0.7511
                            },
                            'siberian': {
                                'ref:long': 0.2727,
                                'alt:long': 0.7273
                            },
                            'tommo': {
                                'ref:long': 0.1747,
                                'alt:long': 0.8253
                            },
                            'topmed': {
                                'ref:long': 0.4822,
                                'alt:long': 0.5178
                            },
                            'twinsuk': {
                                'ref:long': 0.4005,
                                'alt:long': 0.5995
                            },
                            'vietnamese': {
                                'ref:long': 0.1934,
                                'alt:long': 0.8066
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.42,
                                'alt:long': 0.58
                            }
                        },
                        'varinfo': '1-983193-A-G',
                        'vid': '2.46958e+06',
                        'variant_vcf': '1-983193-A-G',
                        'variant_annovar': '1-983193-983193-A-G',
                        'start_position': '983193',
                        'end_position': '983193',
                        'ref_annovar': 'A',
                        'alt_annovar': 'G',
                        'ref_vcf': 'A',
                        'alt_vcf': 'G',
                        'apc_conservation': 0.81362,
                        'apc_conservation_v2': 0.824995,
                        'apc_epigenetics_active': 8.58356,
                        'apc_epigenetics': 11.6437,
                        'apc_epigenetics_repressed': 9.47243,
                        'apc_epigenetics_transcription': 7.52503,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.0384563,
                        'apc_local_nucleotide_diversity_v3': 0.0467769,
                        'apc_mappability': 6.15012,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.0381841,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 14.076,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 9.29793,
                        'apc_transcription_factor': 12.6571,
                        'bravo_an': '264690',
                        'bravo_af': '0.517798',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.31182',
                        'funseq_description': 'noncoding',
                        'af_total': 0.510986,
                        'af_asj_female': 0.572645,
                        'af_eas_female': 0.82663,
                        'af_afr_male': 0.308254,
                        'af_female': 0.506603,
                        'af_fin_male': 0.55823,
                        'af_oth_female': 0.531818,
                        'af_ami': 0.423333,
                        'af_oth': 0.537674,
                        'af_male': 0.515649,
                        'af_ami_female': 0.410638,
                        'af_afr': 0.311564,
                        'af_eas_male': 0.817479,
                        'af_sas': 0.623274,
                        'af_nfe_female': 0.581642,
                        'af_asj_male': 0.59347,
                        'af_raw': 0.510966,
                        'af_oth_male': 0.54381,
                        'af_nfe_male': 0.575727,
                        'af_asj': 0.582431,
                        'af_amr_male': 0.662616,
                        'af_amr_female': 0.640704,
                        'af_fin': 0.5591,
                        'af_afr_female': 0.314388,
                        'af_sas_male': 0.626312,
                        'af_amr': 0.65313,
                        'af_nfe': 0.579153,
                        'af_eas': 0.821703,
                        'af_ami_male': 0.437209,
                        'af_fin_female': 0.561852,
                        'cadd_rawscore': '0.109546',
                        'cadd_phred': '2.906',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.2216',
                        'tg_all': '0.549121',
                        'tg_amr': '0.6455',
                        'tg_eas': '0.8065',
                        'tg_eur': '0.5984',
                        'tg_sas': '0.6074'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 2.30621e-16,
                'slope': -0.807764,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'caf4b2f52b12eac647f51cd9adcadf7aec2af4f0a4ed974a3e91a5c9d2f8cd3f',
                    '_id': 'variants/caf4b2f52b12eac647f51cd9adcadf7aec2af4f0a4ed974a3e91a5c9d2f8cd3f',
                    '_rev': '_geyXJ_S--I',
                    'chr': 'chr1',
                    'pos:long': 983746,
                    'rsid': [
                        'rs28469033'
                    ],
                    'ref': 'T',
                    'alt': 'C',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8339,
                                'alt:long': 0.1661
                            },
                            'alspac': {
                                'ref:long': 0.7818,
                                'alt:long': 0.2182
                            },
                            'estonian': {
                                'ref:long': 0.7438,
                                'alt:long': 0.2562
                            },
                            'genome_dk': {
                                'ref:long': 0.8,
                                'alt:long': 0.2
                            },
                            'gonl': {
                                'ref:long': 0.7766,
                                'alt:long': 0.2234
                            },
                            'korean': {
                                'ref:long': 0.859,
                                'alt:long': 0.141
                            },
                            'korea1k': {
                                'ref:long': 0.8761,
                                'alt:long': 0.1239
                            },
                            'northernsweden': {
                                'ref:long': 0.7183,
                                'alt:long': 0.2817
                            },
                            'qatari': {
                                'ref:long': 0.875,
                                'alt:long': 0.125
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4167,
                                'alt:long': 0.5833
                            },
                            'siberian': {
                                'ref:long': 0.3125,
                                'alt:long': 0.6875
                            },
                            'tommo': {
                                'ref:long': 0.8836,
                                'alt:long': 0.1164
                            },
                            'topmed': {
                                'ref:long': 0.8023,
                                'alt:long': 0.1977
                            },
                            'twinsuk': {
                                'ref:long': 0.771,
                                'alt:long': 0.229
                            },
                            'vietnamese': {
                                'ref:long': 0.9112,
                                'alt:long': 0.08879
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8776,
                                'alt:long': 0.1224
                            }
                        },
                        'varinfo': '1-983747-T-C',
                        'vid': '2.47124e+06',
                        'variant_vcf': '1-983747-T-C',
                        'variant_annovar': '1-983747-983747-T-C',
                        'start_position': '983747',
                        'end_position': '983747',
                        'ref_annovar': 'T',
                        'alt_annovar': 'C',
                        'ref_vcf': 'T',
                        'alt_vcf': 'C',
                        'apc_conservation': 0.0604342,
                        'apc_conservation_v2': 0.0666126,
                        'apc_epigenetics_active': 2.81619,
                        'apc_epigenetics': 4.84861,
                        'apc_epigenetics_repressed': 11.6889,
                        'apc_epigenetics_transcription': 0.749213,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.115484,
                        'apc_local_nucleotide_diversity_v3': 0.1315,
                        'apc_mappability': 4.06326,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.114489,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 4.81849,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 8.50582,
                        'apc_transcription_factor': 9.00041,
                        'bravo_an': '264690',
                        'bravo_af': '0.193517',
                        'filter_status': 'PASS',
                        'fathmm_xf': '4.24319',
                        'funseq_description': 'noncoding',
                        'af_total': 0.202403,
                        'af_asj_female': 0.186364,
                        'af_eas_female': 0.128631,
                        'af_afr_male': 0.161472,
                        'af_female': 0.202426,
                        'af_fin_male': 0.263225,
                        'af_oth_female': 0.186703,
                        'af_ami': 0.270089,
                        'af_oth': 0.191628,
                        'af_male': 0.202378,
                        'af_ami_female': 0.277778,
                        'af_afr': 0.164889,
                        'af_eas_male': 0.136147,
                        'af_sas': 0.169737,
                        'af_nfe_female': 0.22976,
                        'af_asj_male': 0.201024,
                        'af_raw': 0.202533,
                        'af_oth_male': 0.196768,
                        'af_nfe_male': 0.230636,
                        'af_asj': 0.193257,
                        'af_amr_male': 0.165804,
                        'af_amr_female': 0.165481,
                        'af_fin': 0.259854,
                        'af_afr_female': 0.167799,
                        'af_sas_male': 0.169361,
                        'af_amr': 0.165664,
                        'af_nfe': 0.230129,
                        'af_eas': 0.132673,
                        'af_ami_male': 0.261682,
                        'af_fin_female': 0.2492,
                        'cadd_rawscore': '-0.120196',
                        'cadd_phred': '0.891',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1702',
                        'tg_all': '0.166134',
                        'tg_amr': '0.1657',
                        'tg_eas': '0.125',
                        'tg_eur': '0.2068',
                        'tg_sas': '0.1616'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 9.27251e-8,
                'slope': 0.660596,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'b1ea03667bd5a56cc23bf36f0917a38c8616efab8419ed5f00b112d5a5e5d2d8',
                    '_id': 'variants/b1ea03667bd5a56cc23bf36f0917a38c8616efab8419ed5f00b112d5a5e5d2d8',
                    '_rev': '_geyXJDu--B',
                    'chr': 'chr1',
                    'pos:long': 984120,
                    'rsid': [
                        'rs4970414'
                    ],
                    'ref': 'G',
                    'alt': 'T',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4724,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.4139,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.4542,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.475,
                                'alt:long': 0
                            },
                            'gnomad': {
                                'ref:long': 0.5222,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.3002,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.4533,
                                'alt:long': 0
                            },
                            'page_study': {
                                'ref:long': 0.5324,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.5463,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2606,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.3,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.2809,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.5122,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.4067,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4692,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-984121-G-T',
                        'vid': '2.47236e+06',
                        'variant_vcf': '1-984121-G-T',
                        'variant_annovar': '1-984121-984121-G-T',
                        'start_position': '984121',
                        'end_position': '984121',
                        'ref_annovar': 'G',
                        'alt_annovar': 'T',
                        'ref_vcf': 'G',
                        'alt_vcf': 'T',
                        'apc_conservation': 11.076,
                        'apc_conservation_v2': 10.7833,
                        'apc_epigenetics_active': 11.0449,
                        'apc_epigenetics': 14.2076,
                        'apc_epigenetics_repressed': 13.5592,
                        'apc_epigenetics_transcription': 6.4618,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.214541,
                        'apc_local_nucleotide_diversity_v3': 0.224832,
                        'apc_mappability': 5.18561,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.213958,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 0.73334,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 8.10274,
                        'apc_transcription_factor': 18.0097,
                        'bravo_an': '264690',
                        'bravo_af': '0.48785',
                        'filter_status': 'PASS',
                        'fathmm_xf': '3.1944',
                        'funseq_description': 'noncoding',
                        'af_total': 0.481148,
                        'af_asj_female': 0.592614,
                        'af_eas_female': 0.723922,
                        'af_afr_male': 0.24835,
                        'af_female': 0.478807,
                        'af_fin_male': 0.495328,
                        'af_oth_female': 0.52011,
                        'af_ami': 0.528953,
                        'af_oth': 0.510728,
                        'af_male': 0.483642,
                        'af_ami_female': 0.529915,
                        'af_afr': 0.249038,
                        'af_eas_male': 0.714881,
                        'af_sas': 0.637113,
                        'af_nfe_female': 0.575402,
                        'af_asj_male': 0.631748,
                        'af_raw': 0.479975,
                        'af_oth_male': 0.500952,
                        'af_nfe_male': 0.571318,
                        'af_asj': 0.610977,
                        'af_amr_male': 0.610837,
                        'af_amr_female': 0.610564,
                        'af_fin': 0.497984,
                        'af_afr_female': 0.249622,
                        'af_sas_male': 0.639968,
                        'af_amr': 0.610719,
                        'af_nfe': 0.573684,
                        'af_eas': 0.719051,
                        'af_ami_male': 0.527907,
                        'af_fin_female': 0.506405,
                        'cadd_rawscore': '0.71542',
                        'cadd_phred': '9.155',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1732',
                        'tg_all': '0.527556',
                        'tg_amr': '0.6268',
                        'tg_eas': '0.7391',
                        'tg_eur': '0.6064',
                        'tg_sas': '0.637'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 4.94533e-10,
                'slope': -0.601093,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '2bafef001b1e7bb6c12eb79ec305a147a4b5ea8c08e5aeb20dbb6b7b60b52d25',
                    '_id': 'variants/2bafef001b1e7bb6c12eb79ec305a147a4b5ea8c08e5aeb20dbb6b7b60b52d25',
                    '_rev': '_geyXJCG--E',
                    'chr': 'chr1',
                    'pos:long': 985122,
                    'rsid': [
                        'rs28896913'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8852,
                                'alt:long': 0.1148
                            },
                            'alspac': {
                                'ref:long': 0.849,
                                'alt:long': 0.151
                            },
                            'estonian': {
                                'ref:long': 0.8362,
                                'alt:long': 0.1638
                            },
                            'genome_dk': {
                                'ref:long': 0.8,
                                'alt:long': 0.2
                            },
                            'gnomad': {
                                'ref:long': 0.8781,
                                'alt:long': 0.1219
                            },
                            'gonl': {
                                'ref:long': 0.8297,
                                'alt:long': 0.1703
                            },
                            'korean': {
                                'ref:long': 0.9435,
                                'alt:long': 0.05647
                            },
                            'korea1k': {
                                'ref:long': 0.9378,
                                'alt:long': 0.06223
                            },
                            'northernsweden': {
                                'ref:long': 0.8583,
                                'alt:long': 0.1417
                            },
                            'qatari': {
                                'ref:long': 0.7593,
                                'alt:long': 0.2407
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4592,
                                'alt:long': 0.5408
                            },
                            'siberian': {
                                'ref:long': 0.5,
                                'alt:long': 0.5
                            },
                            'tommo': {
                                'ref:long': 0.9171,
                                'alt:long': 0.08294
                            },
                            'topmed': {
                                'ref:long': 0.8823,
                                'alt:long': 0.1177
                            },
                            'twinsuk': {
                                'ref:long': 0.8584,
                                'alt:long': 0.1416
                            },
                            'vietnamese': {
                                'ref:long': 0.9198,
                                'alt:long': 0.08019
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.9221,
                                'alt:long': 0.07793
                            }
                        },
                        'varinfo': '1-985123-G-A',
                        'vid': '2.47537e+06',
                        'variant_vcf': '1-985123-G-A',
                        'variant_annovar': '1-985123-985123-G-A',
                        'start_position': '985123',
                        'end_position': '985123',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 3.49555,
                        'apc_conservation_v2': 3.15033,
                        'apc_epigenetics_active': 3.67682,
                        'apc_epigenetics': 4.37139,
                        'apc_epigenetics_repressed': 3.88354,
                        'apc_epigenetics_transcription': 2.23357,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.348502,
                        'apc_local_nucleotide_diversity_v3': 0.346251,
                        'apc_mappability': 0.666848,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.348064,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 10.3211,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 7.29178,
                        'apc_transcription_factor': 4.69789,
                        'bravo_an': '264690',
                        'bravo_af': '0.117685',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.36041',
                        'funseq_description': 'noncoding',
                        'af_total': 0.122014,
                        'af_asj_female': 0.151705,
                        'af_eas_female': 0.0608575,
                        'af_afr_male': 0.0774728,
                        'af_female': 0.1231,
                        'af_fin_male': 0.141058,
                        'af_oth_female': 0.136612,
                        'af_ami': 0.132517,
                        'af_oth': 0.129198,
                        'af_male': 0.120858,
                        'af_ami_female': 0.134043,
                        'af_afr': 0.0778958,
                        'af_eas_male': 0.0701546,
                        'af_sas': 0.125328,
                        'af_nfe_female': 0.150701,
                        'af_asj_male': 0.141485,
                        'af_raw': 0.122173,
                        'af_oth_male': 0.121415,
                        'af_nfe_male': 0.153291,
                        'af_asj': 0.146899,
                        'af_amr_male': 0.100904,
                        'af_amr_female': 0.108099,
                        'af_fin': 0.144457,
                        'af_afr_female': 0.0782559,
                        'af_sas_male': 0.120467,
                        'af_amr': 0.104017,
                        'af_nfe': 0.151791,
                        'af_eas': 0.0658568,
                        'af_ami_male': 0.130841,
                        'af_fin_female': 0.155227,
                        'cadd_rawscore': '0.136329',
                        'cadd_phred': '3.238',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.087',
                        'tg_all': '0.114816',
                        'tg_amr': '0.1196',
                        'tg_eas': '0.0784',
                        'tg_eur': '0.1561',
                        'tg_sas': '0.1442'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 1.79152e-8,
                'slope': 0.811672,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '5ce175e66a833b4d3fd69d977ec5c6e0b73d45937f1574e656e244b74e4224eb',
                    '_id': 'variants/5ce175e66a833b4d3fd69d977ec5c6e0b73d45937f1574e656e244b74e4224eb',
                    '_rev': '_geyXJAO--Y',
                    'chr': 'chr1',
                    'pos:long': 985259,
                    'rsid': [
                        'rs28869591'
                    ],
                    'ref': 'C',
                    'alt': 'T',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8438,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.8345,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.8132,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.8,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.8257,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.9466,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.94,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.8317,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.7037,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.45,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.5,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.9409,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.8405,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.8436,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8839,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-985260-C-T',
                        'vid': '2.47578e+06',
                        'variant_vcf': '1-985260-C-T',
                        'variant_annovar': '1-985260-985260-C-T',
                        'start_position': '985260',
                        'end_position': '985260',
                        'ref_annovar': 'C',
                        'alt_annovar': 'T',
                        'ref_vcf': 'C',
                        'alt_vcf': 'T',
                        'apc_conservation': 3.75731,
                        'apc_conservation_v2': 3.40874,
                        'apc_epigenetics_active': 6.90641,
                        'apc_epigenetics': 8.67247,
                        'apc_epigenetics_repressed': 11.8978,
                        'apc_epigenetics_transcription': 5.58694,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.284271,
                        'apc_local_nucleotide_diversity_v3': 0.284079,
                        'apc_mappability': 2.26089,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.283977,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 4.22027,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 7.20147,
                        'apc_transcription_factor': 8.85005,
                        'bravo_an': '264690',
                        'bravo_af': '0.159466',
                        'filter_status': 'PASS',
                        'fathmm_xf': '4.65001',
                        'funseq_description': 'noncoding',
                        'af_total': 0.164455,
                        'af_asj_female': 0.181612,
                        'af_eas_female': 0.0642265,
                        'af_afr_male': 0.17019,
                        'af_female': 0.164589,
                        'af_fin_male': 0.181406,
                        'af_oth_female': 0.18,
                        'af_ami': 0.214444,
                        'af_oth': 0.176279,
                        'af_male': 0.164311,
                        'af_ami_female': 0.212766,
                        'af_afr': 0.170846,
                        'af_eas_male': 0.0748219,
                        'af_sas': 0.185648,
                        'af_nfe_female': 0.165481,
                        'af_asj_male': 0.165173,
                        'af_raw': 0.164714,
                        'af_oth_male': 0.172381,
                        'af_nfe_male': 0.169758,
                        'af_asj': 0.173887,
                        'af_amr_male': 0.123288,
                        'af_amr_female': 0.1302,
                        'af_fin': 0.183656,
                        'af_afr_female': 0.171405,
                        'af_sas_male': 0.180016,
                        'af_amr': 0.126282,
                        'af_nfe': 0.167281,
                        'af_eas': 0.0699234,
                        'af_ami_male': 0.216279,
                        'af_fin_female': 0.1908,
                        'cadd_rawscore': '0.257632',
                        'cadd_phred': '4.793',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1853',
                        'tg_all': '0.15615',
                        'tg_amr': '0.1311',
                        'tg_eas': '0.0853',
                        'tg_eur': '0.169',
                        'tg_sas': '0.1943'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 9.94503e-8,
                'slope': 0.741419,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '4298328c6b48b658d3b054d31b8cc917e47da077504a2437d2a5fb9beb0fd561',
                    '_id': 'variants/4298328c6b48b658d3b054d31b8cc917e47da077504a2437d2a5fb9beb0fd561',
                    '_rev': '_geyXJAW--N',
                    'chr': 'chr1',
                    'pos:long': 985596,
                    'rsid': [
                        'rs55692389'
                    ],
                    'ref': 'T',
                    'alt': 'C',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8446,
                                'alt:long': 0.1554
                            },
                            'alspac': {
                                'ref:long': 0.8373,
                                'alt:long': 0.1627
                            },
                            'estonian': {
                                'ref:long': 0.8174,
                                'alt:long': 0.1826
                            },
                            'genome_dk': {
                                'ref:long': 0.85,
                                'alt:long': 0.15
                            },
                            'gnomad': {
                                'ref:long': 0.839,
                                'alt:long': 0.161
                            },
                            'gonl': {
                                'ref:long': 0.8317,
                                'alt:long': 0.1683
                            },
                            'korean': {
                                'ref:long': 0.9466,
                                'alt:long': 0.05339
                            },
                            'korea1k': {
                                'ref:long': 0.94,
                                'alt:long': 0.06004
                            },
                            'northernsweden': {
                                'ref:long': 0.8317,
                                'alt:long': 0.1683
                            },
                            'qatari': {
                                'ref:long': 0.713,
                                'alt:long': 0.287
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4483,
                                'alt:long': 0.5517
                            },
                            'siberian': {
                                'ref:long': 0.5,
                                'alt:long': 0.5
                            },
                            'tommo': {
                                'ref:long': 0.9409,
                                'alt:long': 0.05907
                            },
                            'topmed': {
                                'ref:long': 0.8436,
                                'alt:long': 0.1564
                            },
                            'twinsuk': {
                                'ref:long': 0.8495,
                                'alt:long': 0.1505
                            },
                            'vietnamese': {
                                'ref:long': 0.9019,
                                'alt:long': 0.09813
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8355,
                                'alt:long': 0.1645
                            }
                        },
                        'varinfo': '1-985597-T-C',
                        'vid': '2.47679e+06',
                        'variant_vcf': '1-985597-T-C',
                        'variant_annovar': '1-985597-985597-T-C',
                        'start_position': '985597',
                        'end_position': '985597',
                        'ref_annovar': 'T',
                        'alt_annovar': 'C',
                        'ref_vcf': 'T',
                        'alt_vcf': 'C',
                        'apc_conservation': 2.45435,
                        'apc_conservation_v2': 2.10322,
                        'apc_epigenetics_active': 6.33039,
                        'apc_epigenetics': 5.33823,
                        'apc_epigenetics_repressed': 2.92754,
                        'apc_epigenetics_transcription': 2.29712,
                        'apc_local_nucleotide_diversity': 2.99243,
                        'apc_local_nucleotide_diversity_v2': 0.552782,
                        'apc_local_nucleotide_diversity_v3': 0.498448,
                        'apc_mappability': 2.59819,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.555738,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 2.07265,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.99446,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.156398',
                        'filter_status': 'PASS',
                        'fathmm_xf': '5.82607',
                        'funseq_description': 'noncoding',
                        'af_total': 0.161529,
                        'af_asj_female': 0.177072,
                        'af_eas_female': 0.0630194,
                        'af_afr_male': 0.173001,
                        'af_female': 0.161293,
                        'af_fin_male': 0.180297,
                        'af_oth_female': 0.176364,
                        'af_ami': 0.214444,
                        'af_oth': 0.171309,
                        'af_male': 0.16178,
                        'af_ami_female': 0.212766,
                        'af_afr': 0.173477,
                        'af_eas_male': 0.0753262,
                        'af_sas': 0.184089,
                        'af_nfe_female': 0.158862,
                        'af_asj_male': 0.160051,
                        'af_raw': 0.161721,
                        'af_oth_male': 0.166034,
                        'af_nfe_male': 0.163157,
                        'af_asj': 0.169073,
                        'af_amr_male': 0.120155,
                        'af_amr_female': 0.123816,
                        'af_fin': 0.182775,
                        'af_afr_female': 0.173882,
                        'af_sas_male': 0.178514,
                        'af_amr': 0.12174,
                        'af_nfe': 0.16067,
                        'af_eas': 0.0696486,
                        'af_ami_male': 0.216279,
                        'af_fin_female': 0.190647,
                        'cadd_rawscore': '0.256052',
                        'cadd_phred': '4.774',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1891',
                        'tg_all': '0.155351',
                        'tg_amr': '0.1254',
                        'tg_eas': '0.0853',
                        'tg_eur': '0.165',
                        'tg_sas': '0.1933'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'e15f3a4c5e3fbb33bc0f54b168e16208013dc5b0225392caa123679c4a1eb90e',
                    '_id': 'variants/e15f3a4c5e3fbb33bc0f54b168e16208013dc5b0225392caa123679c4a1eb90e',
                    '_rev': '_geyXJCi--H',
                    'chr': 'chr1',
                    'pos:long': 986279,
                    'rsid': [
                        'rs6703008'
                    ],
                    'ref': 'C',
                    'alt': 'T',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8656,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.7713,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.75,
                                'alt:long': 0
                            },
                            'gonl': {
                                'ref:long': 0.7866,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.8819,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.7267,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.9352,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4589,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.3571,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.8861,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.8311,
                                'alt:long': 0
                            },
                            'vietnamese': {
                                'ref:long': 0.9676,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8996,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-986280-C-T',
                        'vid': '2.47884e+06',
                        'variant_vcf': '1-986280-C-T',
                        'variant_annovar': '1-986280-986280-C-T',
                        'start_position': '986280',
                        'end_position': '986280',
                        'ref_annovar': 'C',
                        'alt_annovar': 'T',
                        'ref_vcf': 'C',
                        'alt_vcf': 'T',
                        'apc_conservation': 0.155207,
                        'apc_conservation_v2': 0.123975,
                        'apc_epigenetics_active': 0.534714,
                        'apc_epigenetics': 0.706612,
                        'apc_epigenetics_repressed': 1.30063,
                        'apc_epigenetics_transcription': 0.543848,
                        'apc_local_nucleotide_diversity': 2.89994,
                        'apc_local_nucleotide_diversity_v2': 0.328978,
                        'apc_local_nucleotide_diversity_v3': 0.323655,
                        'apc_mappability': 0.880723,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.328273,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 1.67399,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.73181,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.168869',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.90487',
                        'funseq_description': 'noncoding',
                        'af_total': 0.178715,
                        'af_asj_female': 0.130114,
                        'af_eas_female': 0.105809,
                        'af_afr_male': 0.11836,
                        'af_female': 0.179137,
                        'af_fin_male': 0.248927,
                        'af_oth_female': 0.165455,
                        'af_ami': 0.257238,
                        'af_oth': 0.168372,
                        'af_male': 0.178266,
                        'af_ami_female': 0.260684,
                        'af_afr': 0.117947,
                        'af_eas_male': 0.0953516,
                        'af_sas': 0.138761,
                        'af_nfe_female': 0.223511,
                        'af_asj_male': 0.134319,
                        'af_raw': 0.178914,
                        'af_oth_male': 0.171429,
                        'af_nfe_male': 0.221604,
                        'af_asj': 0.132087,
                        'af_amr_male': 0.139607,
                        'af_amr_female': 0.144576,
                        'af_fin': 0.244961,
                        'af_afr_female': 0.117595,
                        'af_sas_male': 0.138642,
                        'af_amr': 0.141757,
                        'af_nfe': 0.222708,
                        'af_eas': 0.100192,
                        'af_ami_male': 0.253488,
                        'af_fin_female': 0.232372,
                        'cadd_rawscore': '-0.18374',
                        'cadd_phred': '0.599',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1241',
                        'tg_all': '0.134385',
                        'tg_amr': '0.1542',
                        'tg_eas': '0.0764',
                        'tg_eur': '0.1899',
                        'tg_sas': '0.137'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 6.47548e-7,
                'slope': 0.634734,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'ff45fb4d95bd827f73ec7d08cb2bd28c5bcec222e93bdd2d73009378e34a2997',
                    '_id': 'variants/ff45fb4d95bd827f73ec7d08cb2bd28c5bcec222e93bdd2d73009378e34a2997',
                    '_rev': '_geyXJAi--c',
                    'chr': 'chr1',
                    'pos:long': 986335,
                    'rsid': [
                        'rs13303278'
                    ],
                    'ref': 'C',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.5753,
                                'alt:long': 0.4247
                            },
                            'alspac': {
                                'ref:long': 0.4292,
                                'alt:long': 0.5708
                            },
                            'estonian': {
                                'ref:long': 0.471,
                                'alt:long': 0.529
                            },
                            'genome_dk': {
                                'ref:long': 0.5,
                                'alt:long': 0.5
                            },
                            'gnomad': {
                                'ref:long': 0.58,
                                'alt:long': 0.42
                            },
                            'gonl': {
                                'ref:long': 0.4289,
                                'alt:long': 0.5711
                            },
                            'korean': {
                                'ref:long': 0.4082,
                                'alt:long': 0.5918
                            },
                            'northernsweden': {
                                'ref:long': 0.495,
                                'alt:long': 0.505
                            },
                            'qatari': {
                                'ref:long': 0.6065,
                                'alt:long': 0.3935
                            },
                            'sgdp_prj': {
                                'ref:long': 0.3305,
                                'alt:long': 0.6695
                            },
                            'siberian': {
                                'ref:long': 0.3333,
                                'alt:long': 0.6667
                            },
                            'tommo': {
                                'ref:long': 0.4098,
                                'alt:long': 0.5902
                            },
                            'topmed': {
                                'ref:long': 0.5787,
                                'alt:long': 0.4213
                            },
                            'twinsuk': {
                                'ref:long': 0.4312,
                                'alt:long': 0.5688
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.5153,
                                'alt:long': 0.4847
                            }
                        },
                        'varinfo': '1-986336-C-A',
                        'vid': '2.47901e+06',
                        'variant_vcf': '1-986336-C-A',
                        'variant_annovar': '1-986336-986336-C-A',
                        'start_position': '986336',
                        'end_position': '986336',
                        'ref_annovar': 'C',
                        'alt_annovar': 'A',
                        'ref_vcf': 'C',
                        'alt_vcf': 'A',
                        'apc_conservation': 1.92286,
                        'apc_conservation_v2': 1.62866,
                        'apc_epigenetics_active': 2.19877,
                        'apc_epigenetics': 3.06298,
                        'apc_epigenetics_repressed': 3.05993,
                        'apc_epigenetics_transcription': 0.850371,
                        'apc_local_nucleotide_diversity': 2.89994,
                        'apc_local_nucleotide_diversity_v2': 0.383561,
                        'apc_local_nucleotide_diversity_v3': 0.367246,
                        'apc_mappability': 0.671565,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.383036,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 4.81849,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.72036,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.421293',
                        'filter_status': 'PASS',
                        'fathmm_xf': '3.283',
                        'funseq_description': 'noncoding',
                        'af_total': 0.423076,
                        'af_asj_female': 0.61521,
                        'af_eas_female': 0.583795,
                        'af_afr_male': 0.108149,
                        'af_female': 0.419388,
                        'af_fin_male': 0.514166,
                        'af_oth_female': 0.441606,
                        'af_ami': 0.495556,
                        'af_oth': 0.437966,
                        'af_male': 0.427001,
                        'af_ami_female': 0.485106,
                        'af_afr': 0.112644,
                        'af_eas_male': 0.610514,
                        'af_sas': 0.562377,
                        'af_nfe_female': 0.55825,
                        'af_asj_male': 0.635083,
                        'af_raw': 0.423081,
                        'af_oth_male': 0.43416,
                        'af_nfe_male': 0.556534,
                        'af_asj': 0.624549,
                        'af_amr_male': 0.54643,
                        'af_amr_female': 0.542499,
                        'af_fin': 0.515085,
                        'af_afr_female': 0.116478,
                        'af_sas_male': 0.570105,
                        'af_amr': 0.544728,
                        'af_nfe': 0.557528,
                        'af_eas': 0.59814,
                        'af_ami_male': 0.506977,
                        'af_fin_female': 0.517986,
                        'cadd_rawscore': '0.07139',
                        'cadd_phred': '2.465',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.0318',
                        'tg_all': '0.42472',
                        'tg_amr': '0.5187',
                        'tg_eas': '0.5873',
                        'tg_eur': '0.5815',
                        'tg_sas': '0.5603'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 3.66638e-23,
                'slope': -0.946632,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'f24a1e2049bfd5375f5609a52903ae3176d071fbdf33c9dfd45e330eb4606211',
                    '_id': 'variants/f24a1e2049bfd5375f5609a52903ae3176d071fbdf33c9dfd45e330eb4606211',
                    '_rev': '_geyXJCq--T',
                    'chr': 'chr1',
                    'pos:long': 986628,
                    'rsid': [
                        'rs6693747'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8654,
                                'alt:long': 0.1346
                            },
                            'alspac': {
                                'ref:long': 0.7878,
                                'alt:long': 0.2122
                            },
                            'estonian': {
                                'ref:long': 0.771,
                                'alt:long': 0.229
                            },
                            'genome_dk': {
                                'ref:long': 0.75,
                                'alt:long': 0.25
                            },
                            'gnomad': {
                                'ref:long': 0.8204,
                                'alt:long': 0.1796
                            },
                            'gonl': {
                                'ref:long': 0.7866,
                                'alt:long': 0.2134
                            },
                            'korean': {
                                'ref:long': 0.8816,
                                'alt:long': 0.1184
                            },
                            'northernsweden': {
                                'ref:long': 0.725,
                                'alt:long': 0.275
                            },
                            'qatari': {
                                'ref:long': 0.9352,
                                'alt:long': 0.06481
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4595,
                                'alt:long': 0.5405
                            },
                            'siberian': {
                                'ref:long': 0.3571,
                                'alt:long': 0.6429
                            },
                            'tommo': {
                                'ref:long': 0.8867,
                                'alt:long': 0.1133
                            },
                            'topmed': {
                                'ref:long': 0.8314,
                                'alt:long': 0.1686
                            },
                            'twinsuk': {
                                'ref:long': 0.7762,
                                'alt:long': 0.2238
                            },
                            'vietnamese': {
                                'ref:long': 0.9626,
                                'alt:long': 0.03738
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8006,
                                'alt:long': 0.1994
                            }
                        },
                        'varinfo': '1-986629-G-A',
                        'vid': '2.47988e+06',
                        'variant_vcf': '1-986629-G-A',
                        'variant_annovar': '1-986629-986629-G-A',
                        'start_position': '986629',
                        'end_position': '986629',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 0.151314,
                        'apc_conservation_v2': 0.119085,
                        'apc_epigenetics_active': 7.89911,
                        'apc_epigenetics': 9.08753,
                        'apc_epigenetics_repressed': 19.6131,
                        'apc_epigenetics_transcription': 4.49276,
                        'apc_local_nucleotide_diversity': 2.89994,
                        'apc_local_nucleotide_diversity_v2': 0.480225,
                        'apc_local_nucleotide_diversity_v3': 0.436477,
                        'apc_mappability': 14.511,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.481272,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 14.076,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.66515,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.168552',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.69334',
                        'funseq_description': 'noncoding',
                        'af_total': 0.178719,
                        'af_asj_female': 0.128409,
                        'af_eas_female': 0.106501,
                        'af_afr_male': 0.119768,
                        'af_female': 0.179066,
                        'af_fin_male': 0.248551,
                        'af_oth_female': 0.163024,
                        'af_ami': 0.256667,
                        'af_oth': 0.166512,
                        'af_male': 0.178349,
                        'af_ami_female': 0.259574,
                        'af_afr': 0.118931,
                        'af_eas_male': 0.0969084,
                        'af_sas': 0.137829,
                        'af_nfe_female': 0.223063,
                        'af_asj_male': 0.133333,
                        'af_raw': 0.178881,
                        'af_oth_male': 0.170152,
                        'af_nfe_male': 0.220982,
                        'af_asj': 0.130723,
                        'af_amr_male': 0.139463,
                        'af_amr_female': 0.144505,
                        'af_fin': 0.24497,
                        'af_afr_female': 0.118218,
                        'af_sas_male': 0.138238,
                        'af_amr': 0.141642,
                        'af_nfe': 0.222188,
                        'af_eas': 0.101343,
                        'af_ami_male': 0.253488,
                        'af_fin_female': 0.233626,
                        'cadd_rawscore': '-0.334173',
                        'cadd_phred': '0.201',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1248',
                        'tg_all': '0.134585',
                        'tg_amr': '0.1527',
                        'tg_eas': '0.0764',
                        'tg_eur': '0.1909',
                        'tg_sas': '0.137'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 6.47548e-7,
                'slope': 0.634734,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '9386dd3a5b27f4631a7ad5e464d2e64947ebc71d241c73ffc34251f9940ea231',
                    '_id': 'variants/9386dd3a5b27f4631a7ad5e464d2e64947ebc71d241c73ffc34251f9940ea231',
                    '_rev': '_geyXJCy--U',
                    'chr': 'chr1',
                    'pos:long': 987102,
                    'rsid': [
                        'rs114583033'
                    ],
                    'ref': 'T',
                    'alt': 'C',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.844,
                                'alt:long': 0
                            },
                            'alspac': {
                                'ref:long': 0.8368,
                                'alt:long': 0
                            },
                            'estonian': {
                                'ref:long': 0.8176,
                                'alt:long': 0
                            },
                            'genome_dk': {
                                'ref:long': 0.85,
                                'alt:long': 0
                            },
                            'gnomad': {
                                'ref:long': 0.8403,
                                'alt:long': 0
                            },
                            'korean': {
                                'ref:long': 0.9496,
                                'alt:long': 0
                            },
                            'korea1k': {
                                'ref:long': 0.9432,
                                'alt:long': 0
                            },
                            'northernsweden': {
                                'ref:long': 0.835,
                                'alt:long': 0
                            },
                            'qatari': {
                                'ref:long': 0.713,
                                'alt:long': 0
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4483,
                                'alt:long': 0
                            },
                            'siberian': {
                                'ref:long': 0.5,
                                'alt:long': 0
                            },
                            'tommo': {
                                'ref:long': 0.9431,
                                'alt:long': 0
                            },
                            'topmed': {
                                'ref:long': 0.8432,
                                'alt:long': 0
                            },
                            'twinsuk': {
                                'ref:long': 0.8501,
                                'alt:long': 0
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8428,
                                'alt:long': 0
                            }
                        },
                        'varinfo': '1-987103-T-C',
                        'vid': '2.48131e+06',
                        'variant_vcf': '1-987103-T-C',
                        'variant_annovar': '1-987103-987103-T-C',
                        'start_position': '987103',
                        'end_position': '987103',
                        'ref_annovar': 'T',
                        'alt_annovar': 'C',
                        'ref_vcf': 'T',
                        'alt_vcf': 'C',
                        'apc_conservation': 0.0947274,
                        'apc_conservation_v2': 0.0755336,
                        'apc_epigenetics_active': 0.75796,
                        'apc_epigenetics': 1.20067,
                        'apc_epigenetics_repressed': 2.21873,
                        'apc_epigenetics_transcription': 2.08839,
                        'apc_local_nucleotide_diversity': 2.81345,
                        'apc_local_nucleotide_diversity_v2': 4.02152,
                        'apc_local_nucleotide_diversity_v3': 2.58412,
                        'apc_mappability': 1.94868,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 4.21616,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 0.205447,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.59197,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.156783',
                        'filter_status': 'PASS',
                        'fathmm_xf': '2.74681',
                        'funseq_description': 'noncoding',
                        'af_total': 0.160098,
                        'af_asj_female': 0.175818,
                        'af_eas_female': 0.0619534,
                        'af_afr_male': 0.172898,
                        'af_female': 0.159863,
                        'af_fin_male': 0.177006,
                        'af_oth_female': 0.179563,
                        'af_ami': 0.213948,
                        'af_oth': 0.170051,
                        'af_male': 0.16035,
                        'af_ami_female': 0.204955,
                        'af_afr': 0.172974,
                        'af_eas_male': 0.0715182,
                        'af_sas': 0.180377,
                        'af_nfe_female': 0.157852,
                        'af_asj_male': 0.156332,
                        'af_raw': 0.163866,
                        'af_oth_male': 0.160083,
                        'af_nfe_male': 0.162516,
                        'af_asj': 0.166667,
                        'af_amr_male': 0.118715,
                        'af_amr_female': 0.122234,
                        'af_fin': 0.178498,
                        'af_afr_female': 0.173038,
                        'af_sas_male': 0.17363,
                        'af_amr': 0.120246,
                        'af_nfe': 0.159816,
                        'af_eas': 0.0670937,
                        'af_ami_male': 0.223881,
                        'af_fin_female': 0.1835,
                        'cadd_rawscore': '-0.062211',
                        'cadd_phred': '1.25',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1899',
                        'tg_all': '0.15595',
                        'tg_amr': '0.1254',
                        'tg_eas': '0.0853',
                        'tg_eur': '0.167',
                        'tg_sas': '0.1933'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '14e9c0b218f3cbd186fe91800d24caf4bd77459110d2897f45be6da709408cc6',
                    '_id': 'variants/14e9c0b218f3cbd186fe91800d24caf4bd77459110d2897f45be6da709408cc6',
                    '_rev': '_geyXJK---B',
                    'chr': 'chr1',
                    'pos:long': 989067,
                    'rsid': [
                        'rs141818071'
                    ],
                    'ref': 'GTTGA',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.4613,
                                'alt:long': 0.5387
                            },
                            'alspac': {
                                'ref:long': 0.4128,
                                'alt:long': 0.5872
                            },
                            'genome_dk': {
                                'ref:long': 0.45,
                                'alt:long': 0.55
                            },
                            'gnomad': {
                                'ref:long': 0.5132,
                                'alt:long': 0.4868
                            },
                            'gonl': {
                                'ref:long': 0.4148,
                                'alt:long': 0.5852
                            },
                            'korea1k': {
                                'ref:long': 0.209,
                                'alt:long': 0.791
                            },
                            'northernsweden': {
                                'ref:long': 0.4932,
                                'alt:long': 0.5068
                            },
                            'tommo': {
                                'ref:long': 0.2396,
                                'alt:long': 0.7604
                            },
                            'twinsuk': {
                                'ref:long': 0.4161,
                                'alt:long': 0.5839
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.4835,
                                'alt:long': 0.5165
                            }
                        },
                        'varinfo': '1-989068-GTTGA-G',
                        'vid': '8.81294e+09',
                        'variant_vcf': '1-989068-GTTGA-G',
                        'variant_annovar': '1-989069-989072-TTGA--',
                        'start_position': '989069',
                        'end_position': '989072',
                        'ref_annovar': 'TTGA',
                        'alt_annovar': '-',
                        'ref_vcf': 'GTTGA',
                        'alt_vcf': 'G',
                        'bravo_an': '264690',
                        'bravo_af': '0.498553',
                        'filter_status': 'SVM',
                        'af_total': 0.489959,
                        'af_asj_female': 0.621404,
                        'af_eas_female': 0.756429,
                        'af_afr_male': 0.256684,
                        'af_female': 0.487406,
                        'af_fin_male': 0.512712,
                        'af_oth_female': 0.517891,
                        'af_ami': 0.502288,
                        'af_oth': 0.521215,
                        'af_male': 0.49271,
                        'af_ami_female': 0.493333,
                        'af_afr': 0.263605,
                        'af_eas_male': 0.763421,
                        'af_sas': 0.639559,
                        'af_nfe_female': 0.571198,
                        'af_asj_male': 0.647251,
                        'af_raw': 0.493732,
                        'af_oth_male': 0.524704,
                        'af_nfe_male': 0.567213,
                        'af_asj': 0.633497,
                        'af_amr_male': 0.660044,
                        'af_amr_female': 0.648135,
                        'af_fin': 0.517267,
                        'af_afr_female': 0.269501,
                        'af_sas_male': 0.644848,
                        'af_amr': 0.654877,
                        'af_nfe': 0.569523,
                        'af_eas': 0.76016,
                        'af_ami_male': 0.511792,
                        'af_fin_female': 0.530604,
                        'cadd_rawscore': '-0.101198',
                        'cadd_phred': '0.998',
                        'tg_afr': '0.1997',
                        'tg_all': '0.538738',
                        'tg_amr': '0.6297',
                        'tg_eas': '0.7708',
                        'tg_eur': '0.5934',
                        'tg_sas': '0.637'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 2.03398e-22,
                'slope': -0.899991,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '75a7f16eea568ee3fb51bfa17d5880c22679ef21986e9a9812a290ee0d994912',
                    '_id': 'variants/75a7f16eea568ee3fb51bfa17d5880c22679ef21986e9a9812a290ee0d994912',
                    '_rev': '_geyXJBq--M',
                    'chr': 'chr1',
                    'pos:long': 989147,
                    'rsid': [
                        'rs34712273'
                    ],
                    'ref': 'C',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.5148,
                                'alt:long': 0.4852
                            },
                            'alspac': {
                                'ref:long': 0.4144,
                                'alt:long': 0.5856
                            },
                            'genome_dk': {
                                'ref:long': 0.45,
                                'alt:long': 0.55
                            },
                            'gonl': {
                                'ref:long': 0.4148,
                                'alt:long': 0.5852
                            },
                            'korean': {
                                'ref:long': 0.2247,
                                'alt:long': 0.7753
                            },
                            'korea1k': {
                                'ref:long': 0.206,
                                'alt:long': 0.794
                            },
                            'northernsweden': {
                                'ref:long': 0.485,
                                'alt:long': 0.515
                            },
                            'qatari': {
                                'ref:long': 0.6065,
                                'alt:long': 0.3935
                            },
                            'sgdp_prj': {
                                'ref:long': 0.2338,
                                'alt:long': 0.7616
                            },
                            'siberian': {
                                'ref:long': 0.3125,
                                'alt:long': 0.6875
                            },
                            'tommo': {
                                'ref:long': 0.2738,
                                'alt:long': 0.7262
                            },
                            'twinsuk': {
                                'ref:long': 0.4191,
                                'alt:long': 0.5809
                            },
                            'vietnamese': {
                                'ref:long': 0.2379,
                                'alt:long': 0.7621
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.6991,
                                'alt:long': 0.3009
                            }
                        },
                        'varinfo': '1-989148-C-A',
                        'vid': '2.48744e+06',
                        'variant_vcf': '1-989148-C-A',
                        'variant_annovar': '1-989148-989148-C-A',
                        'start_position': '989148',
                        'end_position': '989148',
                        'ref_annovar': 'C',
                        'alt_annovar': 'A',
                        'ref_vcf': 'C',
                        'alt_vcf': 'A',
                        'apc_conservation': 6.62257,
                        'apc_conservation_v2': 5.63627,
                        'apc_epigenetics_active': 2.34603,
                        'apc_epigenetics': 3.05526,
                        'apc_epigenetics_repressed': 5.82915,
                        'apc_epigenetics_transcription': 2.61187,
                        'apc_local_nucleotide_diversity': 2.37776,
                        'apc_local_nucleotide_diversity_v2': 8.4289,
                        'apc_local_nucleotide_diversity_v3': 5.50453,
                        'apc_mappability': 1.35858,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 8.86244,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 0.678561,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 6.46144,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.438913',
                        'filter_status': 'SVM',
                        'fathmm_xf': '5.33901',
                        'funseq_description': 'noncoding',
                        'af_total': 0.431503,
                        'af_asj_female': 0.611702,
                        'af_eas_female': 0.741124,
                        'af_afr_male': 0.10822,
                        'af_female': 0.425434,
                        'af_fin_male': 0.511013,
                        'af_oth_female': 0.460728,
                        'af_ami': 0.489183,
                        'af_oth': 0.46523,
                        'af_male': 0.438009,
                        'af_ami_female': 0.493274,
                        'af_afr': 0.112534,
                        'af_eas_male': 0.754068,
                        'af_sas': 0.625174,
                        'af_nfe_female': 0.558001,
                        'af_asj_male': 0.632667,
                        'af_raw': 0.438876,
                        'af_oth_male': 0.46994,
                        'af_nfe_male': 0.556429,
                        'af_asj': 0.621554,
                        'af_amr_male': 0.62706,
                        'af_amr_female': 0.612491,
                        'af_fin': 0.513785,
                        'af_afr_female': 0.116203,
                        'af_sas_male': 0.629392,
                        'af_amr': 0.620754,
                        'af_nfe': 0.55734,
                        'af_eas': 0.748136,
                        'af_ami_male': 0.484456,
                        'af_fin_female': 0.522232,
                        'cadd_rawscore': '0.309932',
                        'cadd_phred': '5.414',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.0348',
                        'tg_all': '0.485224',
                        'tg_amr': '0.6153',
                        'tg_eas': '0.7639',
                        'tg_eur': '0.5785',
                        'tg_sas': '0.6186'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 1.27736e-20,
                'slope': -0.925848,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': 'd2951bddb245eba65291f34d65c487c180940a923a5f93769ed3c06d514966c5',
                    '_id': 'variants/d2951bddb245eba65291f34d65c487c180940a923a5f93769ed3c06d514966c5',
                    '_rev': '_geyXJWW--c',
                    'chr': 'chr1',
                    'pos:long': 991587,
                    'rsid': [
                        'rs140144962'
                    ],
                    'ref': 'C',
                    'alt': 'T',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8413,
                                'alt:long': 0.1587
                            },
                            'alspac': {
                                'ref:long': 0.8371,
                                'alt:long': 0.1629
                            },
                            'estonian': {
                                'ref:long': 0.8182,
                                'alt:long': 0.1818
                            },
                            'gnomad': {
                                'ref:long': 0.8354,
                                'alt:long': 0.1646
                            },
                            'gonl': {
                                'ref:long': 0.8327,
                                'alt:long': 0.1673
                            },
                            'korean': {
                                'ref:long': 0.9497,
                                'alt:long': 0.05031
                            },
                            'korea1k': {
                                'ref:long': 0.9432,
                                'alt:long': 0.05677
                            },
                            'northernsweden': {
                                'ref:long': 0.8317,
                                'alt:long': 0.1683
                            },
                            'qatari': {
                                'ref:long': 0.7222,
                                'alt:long': 0.2778
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4483,
                                'alt:long': 0.5517
                            },
                            'siberian': {
                                'ref:long': 0.5,
                                'alt:long': 0.5
                            },
                            'tommo': {
                                'ref:long': 0.9424,
                                'alt:long': 0.05758
                            },
                            'topmed': {
                                'ref:long': 0.8408,
                                'alt:long': 0.1592
                            },
                            'twinsuk': {
                                'ref:long': 0.8495,
                                'alt:long': 0.1505
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8343,
                                'alt:long': 0.1657
                            }
                        },
                        'varinfo': '1-991588-C-T',
                        'vid': '2.49476e+06',
                        'variant_vcf': '1-991588-C-T',
                        'variant_annovar': '1-991588-991588-C-T',
                        'start_position': '991588',
                        'end_position': '991588',
                        'ref_annovar': 'C',
                        'alt_annovar': 'T',
                        'ref_vcf': 'C',
                        'alt_vcf': 'T',
                        'apc_conservation': 0.124635,
                        'apc_conservation_v2': 0.0981941,
                        'apc_epigenetics_active': 2.80019,
                        'apc_epigenetics': 2.37955,
                        'apc_epigenetics_repressed': 2.55547,
                        'apc_epigenetics_transcription': 0.525597,
                        'apc_local_nucleotide_diversity': 2.13098,
                        'apc_local_nucleotide_diversity_v2': 2.15386,
                        'apc_local_nucleotide_diversity_v3': 1.407,
                        'apc_mappability': 0.506935,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 2.25438,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 12.2324,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 7.2423,
                        'apc_transcription_factor': 3.14279,
                        'bravo_an': '264690',
                        'bravo_af': '0.159216',
                        'filter_status': 'PASS',
                        'fathmm_xf': '3.02062',
                        'funseq_description': 'noncoding',
                        'af_total': 0.165056,
                        'af_asj_female': 0.177107,
                        'af_eas_female': 0.0630252,
                        'af_afr_male': 0.183457,
                        'af_female': 0.164789,
                        'af_fin_male': 0.181366,
                        'af_oth_female': 0.17792,
                        'af_ami': 0.213647,
                        'af_oth': 0.173364,
                        'af_male': 0.16534,
                        'af_ami_female': 0.213675,
                        'af_afr': 0.183508,
                        'af_eas_male': 0.0717703,
                        'af_sas': 0.186458,
                        'af_nfe_female': 0.159723,
                        'af_asj_male': 0.159383,
                        'af_raw': 0.164686,
                        'af_oth_male': 0.168582,
                        'af_nfe_male': 0.164324,
                        'af_asj': 0.16878,
                        'af_amr_male': 0.121066,
                        'af_amr_female': 0.12406,
                        'af_fin': 0.183717,
                        'af_afr_female': 0.183552,
                        'af_sas_male': 0.18059,
                        'af_amr': 0.12236,
                        'af_nfe': 0.161659,
                        'af_eas': 0.0677419,
                        'af_ami_male': 0.213615,
                        'af_fin_female': 0.191106,
                        'cadd_rawscore': '-0.234736',
                        'cadd_phred': '0.424',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.2005',
                        'tg_all': '0.158746',
                        'tg_amr': '0.1254',
                        'tg_eas': '0.0843',
                        'tg_eur': '0.166',
                        'tg_sas': '0.1953'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 7.46515e-8,
                'slope': 0.761218,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '016b278013174d0a810967181cd46decdf6d040c0a97a5e096d4983a192b3f5b',
                    '_id': 'variants/016b278013174d0a810967181cd46decdf6d040c0a97a5e096d4983a192b3f5b',
                    '_rev': '_geyXJWq--_',
                    'chr': 'chr1',
                    'pos:long': 992360,
                    'rsid': [
                        'rs6665587'
                    ],
                    'ref': 'G',
                    'alt': 'A',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            '1000genomes': {
                                'ref:long': 0.8624,
                                'alt:long': 0.1376
                            },
                            'estonian': {
                                'ref:long': 0.7701,
                                'alt:long': 0.2299
                            },
                            'genome_dk': {
                                'ref:long': 0.75,
                                'alt:long': 0.25
                            },
                            'gnomad': {
                                'ref:long': 0.8183,
                                'alt:long': 0.1817
                            },
                            'gonl': {
                                'ref:long': 0.7836,
                                'alt:long': 0.2164
                            },
                            'korean': {
                                'ref:long': 0.8823,
                                'alt:long': 0.1177
                            },
                            'northernsweden': {
                                'ref:long': 0.725,
                                'alt:long': 0.275
                            },
                            'qatari': {
                                'ref:long': 0.9167,
                                'alt:long': 0.08333
                            },
                            'sgdp_prj': {
                                'ref:long': 0.4527,
                                'alt:long': 0.5473
                            },
                            'siberian': {
                                'ref:long': 0.3571,
                                'alt:long': 0.6429
                            },
                            'tommo': {
                                'ref:long': 0.8868,
                                'alt:long': 0.1132
                            },
                            'topmed': {
                                'ref:long': 0.8291,
                                'alt:long': 0.1709
                            },
                            'vietnamese': {
                                'ref:long': 0.9626,
                                'alt:long': 0.03738
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.8043,
                                'alt:long': 0.1957
                            }
                        },
                        'varinfo': '1-992361-G-A',
                        'vid': '2.49708e+06',
                        'variant_vcf': '1-992361-G-A',
                        'variant_annovar': '1-992361-992361-G-A',
                        'start_position': '992361',
                        'end_position': '992361',
                        'ref_annovar': 'G',
                        'alt_annovar': 'A',
                        'ref_vcf': 'G',
                        'alt_vcf': 'A',
                        'apc_conservation': 0.149279,
                        'apc_conservation_v2': 0.117828,
                        'apc_epigenetics_active': 1.1854,
                        'apc_epigenetics': 1.36925,
                        'apc_epigenetics_repressed': 3.02482,
                        'apc_epigenetics_transcription': 0.640447,
                        'apc_local_nucleotide_diversity': 2.03356,
                        'apc_local_nucleotide_diversity_v2': 0.751954,
                        'apc_local_nucleotide_diversity_v3': 0.612692,
                        'apc_mappability': 0.690225,
                        'apc_micro_rna': 99.4512,
                        'apc_mutation_density': 0.772497,
                        'apc_protein_function': 20.2494,
                        'apc_protein_function_v2': 4.92793e-10,
                        'apc_protein_function_v3': 2.96949,
                        'apc_proximity_to_coding': 6.38527,
                        'apc_proximity_to_coding_v2': 15.1786,
                        'apc_proximity_to_tsstes': 7.7361,
                        'apc_transcription_factor': 4.69789,
                        'bravo_an': '264690',
                        'bravo_af': '0.170943',
                        'filter_status': 'PASS',
                        'fathmm_xf': '3.27321',
                        'funseq_description': 'noncoding',
                        'af_total': 0.18107,
                        'af_asj_female': 0.128977,
                        'af_eas_female': 0.105263,
                        'af_afr_male': 0.1196,
                        'af_female': 0.181381,
                        'af_fin_male': 0.251202,
                        'af_oth_female': 0.166364,
                        'af_ami': 0.256667,
                        'af_oth': 0.171163,
                        'af_male': 0.180739,
                        'af_ami_female': 0.259574,
                        'af_afr': 0.118817,
                        'af_eas_male': 0.0988095,
                        'af_sas': 0.151415,
                        'af_nfe_female': 0.227314,
                        'af_asj_male': 0.133333,
                        'af_raw': 0.181129,
                        'af_oth_male': 0.17619,
                        'af_nfe_male': 0.225208,
                        'af_asj': 0.131024,
                        'af_amr_male': 0.138595,
                        'af_amr_female': 0.144407,
                        'af_fin': 0.24726,
                        'af_afr_female': 0.118151,
                        'af_sas_male': 0.151172,
                        'af_amr': 0.141108,
                        'af_nfe': 0.226427,
                        'af_eas': 0.101793,
                        'af_ami_male': 0.253488,
                        'af_fin_female': 0.2348,
                        'cadd_rawscore': '-0.269039',
                        'cadd_phred': '0.331',
                        'refseq_category': 'intergenic',
                        'tg_afr': '0.1241',
                        'tg_all': '0.13758',
                        'tg_amr': '0.1527',
                        'tg_eas': '0.0764',
                        'tg_eur': '0.1938',
                        'tg_sas': '0.1503'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 3.89225e-7,
                'slope': 0.644408,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            },
            {
                'sequence variant': {
                    '_key': '3be794b104906c9257577eb766f68df91302a58f380c6f2b14e40a8bdfcf31e3',
                    '_id': 'variants/3be794b104906c9257577eb766f68df91302a58f380c6f2b14e40a8bdfcf31e3',
                    '_rev': '_geyXJW2--O',
                    'chr': 'chr1',
                    'pos:long': 992966,
                    'rsid': [
                        'rs760413453'
                    ],
                    'ref': 'GGGAGGGTCCATGTGTCCGTCATCTGA',
                    'alt': 'G',
                    'qual': '.',
                    'filter': None,
                    'annotations': {
                        'freq': {
                            'genome_dk': {
                                'ref:long': 0.6,
                                'alt:long': 0.4
                            },
                            'gnomad': {
                                'ref:long': 0.573,
                                'alt:long': 0.427
                            },
                            'tommo': {
                                'ref:long': 0.2497,
                                'alt:long': 0.7503
                            },
                            'dbgap_popfreq': {
                                'ref:long': 0.7511,
                                'alt:long': 0.2489
                            }
                        },
                        'varinfo': '1-992967-GGGAGGGTCCATGTGTCCGTCATCTGA-G',
                        'vid': '8.81294e+09',
                        'variant_vcf': '1-992967-GGGAGGGTCCATGTGTCCGTCATCTGA-G',
                        'variant_annovar': '1-992968-992993-GGAGGGTCCATGTGTCCGTCATCTGA--',
                        'start_position': '992968',
                        'end_position': '992993',
                        'ref_annovar': 'GGAGGGTCCATGTGTCCGTCATCTGA',
                        'alt_annovar': '-',
                        'ref_vcf': 'GGGAGGGTCCATGTGTCCGTCATCTGA',
                        'alt_vcf': 'G',
                        'bravo_an': '264690',
                        'bravo_af': '0.170592',
                        'filter_status': 'SVM',
                        'af_total': 0.430212,
                        'af_asj_female': 0.60943,
                        'af_eas_female': 0.740469,
                        'af_afr_male': 0.116444,
                        'af_female': 0.427655,
                        'af_fin_male': 0.486909,
                        'af_oth_female': 0.45648,
                        'af_ami': 0.485915,
                        'af_oth': 0.456306,
                        'af_male': 0.432991,
                        'af_ami_female': 0.480088,
                        'af_afr': 0.12002,
                        'af_eas_male': 0.747442,
                        'af_sas': 0.596595,
                        'af_nfe_female': 0.553145,
                        'af_asj_male': 0.637022,
                        'af_raw': 0.435253,
                        'af_oth_male': 0.456122,
                        'af_nfe_male': 0.546087,
                        'af_asj': 0.622373,
                        'af_amr_male': 0.626628,
                        'af_amr_female': 0.61568,
                        'af_fin': 0.496255,
                        'af_afr_female': 0.123075,
                        'af_sas_male': 0.601563,
                        'af_amr': 0.621907,
                        'af_nfe': 0.550192,
                        'af_eas': 0.744194,
                        'af_ami_male': 0.4925,
                        'af_fin_female': 0.52292,
                        'cadd_rawscore': '0.736549',
                        'cadd_phred': '9.337'
                    },
                    'format': '.',
                    'source': 'FAVOR',
                    'source_url': 'http://favor.genohub.org/'
                },
                'gene': {
                    '_key': 'ENSG00000188290',
                    '_id': 'genes/ENSG00000188290',
                    '_rev': '_gW3Pv9G---',
                    'gene_id': 'ENSG00000188290.11',
                    'gene_type': 'protein_coding',
                    'chr': 'chr1',
                    'start': 998961,
                    'end': 1000172,
                    'gene_name': 'HES4',
                    'source': 'GENCODE',
                    'version': 'v43',
                    'source_url': 'https://www.gencodegenes.org/human/',
                    'alias': [
                        'HES4',
                        'class B basic helix-loop-helix protein 42',
                        'transcription factor HES-4',
                        'hes family bHLH transcription factor 4',
                        'HGNC:24149',
                        'hairy and enhancer of split 4',
                        'hHES4',
                        'bHLH factor Hes4',
                        'bHLHb42'
                    ],
                    'end:long': 1000172,
                    'start:long': 998961
                },
                'p_value': 2.3738e-24,
                'slope': -0.928406,
                'beta': 3.05583e-14,
                'label': 'splice_QTL',
                'source': 'GTEx',
                'biological_context': 'Brain_Nucleus_accumbens_basal_ganglia',
                'intron_chr': 'chr1',
                'intron_start': 999787,
                'intron_end': 999866
            }
        ]
    }]
}
